{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **CHAPTER 2. Phylogenomic analysis**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import all the modules needed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import pandas as pd\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create the directory to store the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "mkdir phylogenomics/\n",
    "mkdir phylogenomics/Proteins_renamed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Take sequences files from `pangenome/Annotation/Proteins_classic` directory and rename them all to `>[GENE]` format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "for file in pangenome/Annotation/Proteins_classic/*.prt\n",
    "do \n",
    "  sed -E -e 's/^.*(ND1|nd1|NAD1|nad1).*/>NAD1/' \\\n",
    "         -e 's/^.*(ND2|nd2|NAD2|nad2).*/>NAD2/' \\\n",
    "         -e 's/^.*(ND3|nd3|NAD3|nad3).*/>NAD3/' \\\n",
    "         -e 's/^.*(ND4|nd4|NAD4|nad4).*/>NAD4/' \\\n",
    "         -e 's/^.*(ND4L|nd4l|NAD4L|nad4l|NAD4l|nad4L).*/>NAD4L/' \\\n",
    "         -e 's/^.*(ND5|nd5|NAD5|nad5).*/>NAD5/' \\\n",
    "         -e 's/^.*(ND6|nd6|NAD6|nad6).*/>NAD6/' \\\n",
    "         -e 's/^.*(COX1|cox1).*/>COX1/' \\\n",
    "         -e 's/^.*(COX2|cox2).*/>COX2/' \\\n",
    "         -e 's/^.*(COX3|cox3).*/>COX3/' \\\n",
    "         -e 's/^.*(COB|cob).*/>COB/' \\\n",
    "         -e 's/^.*(rps3|RPS3).*/>rps3/' \\\n",
    "         -e 's/^.*(ATP6|atp6).*/>ATP6/' \\\n",
    "         -e 's/^.*(ATP8|atp8).*/>ATP8/' \\\n",
    "         -e 's/^.*(ATP9|atp9).*/>ATP9/' \\\n",
    "         -e 's/^.*(CYTB|cytb).*/>CYTB/' \"$file\" | \\\n",
    "  awk '/^>/ {keep=0} /^>(NAD1|NAD2|NAD3|NAD4|NAD4L|NAD5|NAD6|COX1|COX2|COX3|COB|rps3|ATP6|ATP8|ATP9|CYTB)/ {keep=1} keep' \\\n",
    "  > phylogenomics/Proteins_renamed/$(basename \"$file\" .prt)_renamed_mt_prot.fa\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">COX1\n"
     ]
    }
   ],
   "source": [
    "! head -1 phylogenomics/Proteins_renamed/NC_015789_renamed_mt_prot.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Good."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Download `BBMap`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! wget https://sourceforge.net/projects/bbmap/files/latest/download -O BBMap.tar.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Unarchive `BBMap` and delete the initial archive"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! tar -xvzf BBMap.tar.gz && rm -rf BBMap.tar.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a directory to store renamed sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! mkdir phylogenomics/Proteins_renamed_r2/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using `BBMap` rename sequences from `>[GENE]` format to `>accession_number_[GENE]` format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "for file in phylogenomics/Proteins_renamed/*mt_prot.fa\n",
    "do export species_name=$(basename $file _renamed_mt_prot.fa)\n",
    "bbmap/rename.sh in=$file prefix=\"$species_name\" addprefix=true out=phylogenomics/Proteins_renamed_r2/$species_name.mt_prots.fa ignorejunk=true ;\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">NC_015789_COX1\n"
     ]
    }
   ],
   "source": [
    "! head -1 phylogenomics/Proteins_renamed_r2/NC_015789.mt_prots.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Good."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some genes could be not single copy genes. In some organisms there may be several `cox1` genes. That's why we need to enumerate them!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rename_sequences(input_dir, output_dir):\n",
    "    # Create the output directory if it doesn't exist\n",
    "    os.makedirs(output_dir, exist_ok=True)\n",
    "    \n",
    "    # Loop through each FASTA file in the directory\n",
    "    for filename in os.listdir(input_dir):\n",
    "        if filename.endswith(\".fasta\") or filename.endswith(\".fa\"):\n",
    "            filepath = os.path.join(input_dir, filename)\n",
    "            output_filepath = os.path.join(output_dir, filename)\n",
    "\n",
    "            # Dictionary to keep track of sequence names and their counts\n",
    "            name_count = {}\n",
    "            \n",
    "            # List to store the updated sequences\n",
    "            updated_sequences = []\n",
    "            \n",
    "            # Read the FASTA file\n",
    "            for record in SeqIO.parse(filepath, \"fasta\"):\n",
    "                name = record.id\n",
    "                \n",
    "                # If the name is already seen, append a count suffix\n",
    "                if name in name_count:\n",
    "                    name_count[name] += 1\n",
    "                    new_name = f\"{name}_{name_count[name]}\"\n",
    "                else:\n",
    "                    name_count[name] = 1\n",
    "                    new_name = f\"{name}_1\"\n",
    "                \n",
    "                # Update the record ID\n",
    "                record.id = new_name\n",
    "                record.description = \"\"  # Optionally clear the description\n",
    "                \n",
    "                # Store the updated record\n",
    "                updated_sequences.append(record)\n",
    "\n",
    "            # Write the updated sequences to a new FASTA file\n",
    "            with open(output_filepath, \"w\") as output_handle:\n",
    "                SeqIO.write(updated_sequences, output_handle, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dir = \"phylogenomics/Proteins_renamed_r2\"\n",
    "output_dir = \"phylogenomics/Proteins_renamed_r3\"\n",
    "\n",
    "rename_sequences(input_dir, output_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">NC_015789_COX1_1\n"
     ]
    }
   ],
   "source": [
    "! head -1 phylogenomics/Proteins_renamed_r3/NC_015789.mt_prots.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Good."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we shall prepare to launch `Proteinortho`. `Proteinortho` has a lot of requirements for inputs. For instance, if in directory there are empty files - `Proteinortho` will just crash. So, we delete empty files!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "for file in os.listdir(\"phylogenomics/Proteins_renamed_r3\"):\n",
    "    path = os.path.join(\"phylogenomics/Proteins_renamed_r3\", file)\n",
    "    if os.path.isfile(path) and os.path.getsize(path) == 0:\n",
    "        os.remove(path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Launch `Proteinortho`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*****************************************************************\n",
      "\u001b[1;32mProteinortho\u001b[0m with PoFF version 6.3.4 - An orthology detection tool\n",
      "*****************************************************************\n",
      "Using 10 CPU threads (1 threads per processes each with 10 threads), Detected 'diamond' version 2.1.11\n",
      "Checking input files.\n",
      "Checking phylogenomics/Proteins_renamed_r3/NC_015789.mt_prots.fa... ok\n",
      "Checking phylogenomics/Proteins_renamed_r3/NC_015991.mt_prots.fa... ok\n",
      "Checking phylogenomics/Proteins_renamed_r3/NC_023125.mt_prots.fa... ok\n",
      "Checking phylogenomics/Proteins_renamed_r3/NC_023126.mt_prots.fa... ok\n",
      "Checking phylogenomics/Proteins_renamed_r3/NC_023127.mt_prots.fa... ok\n",
      "Checking phylogenomics/Proteins_renamed_r3/NC_023128.mt_prots.fa... ok\n",
      "Checking phylogenomics/Proteins_renamed_r3/NC_025200.mt_prots.fa... ok\n",
      "Checking phylogenomics/Proteins_renamed_r3/NC_027422.mt_prots.fa... ok\n",
      "Checking phylogenomics/Proteins_renamed_r3/NC_031375.mt_prots.fa... ok\n",
      "Checking phylogenomics/Proteins_renamed_r3/NC_033907.mt_prots.fa... ok\n",
      "Checking phylogenomics/Proteins_renamed_r3/NC_048507.mt_prots.fa... ok\n",
      "Checking phylogenomics/Proteins_renamed_r3/NC_051483.mt_prots.fa... ok\n",
      "Checking phylogenomics/Proteins_renamed_r3/NC_056146.mt_prots.fa... ok\n",
      "Checking phylogenomics/Proteins_renamed_r3/NC_056147.mt_prots.fa... ok\n",
      "Checking phylogenomics/Proteins_renamed_r3/NC_056148.mt_prots.fa... ok\n",
      "Checking phylogenomics/Proteins_renamed_r3/NC_056195.mt_prots.fa... ok\n",
      "Checking phylogenomics/Proteins_renamed_r3/NC_061762.mt_prots.fa... ok\n",
      "Checking phylogenomics/Proteins_renamed_r3/NC_069201.mt_prots.fa... ok\n",
      "Checking phylogenomics/Proteins_renamed_r3/NC_070174.mt_prots.fa... ok\n",
      "Checking phylogenomics/Proteins_renamed_r3/NC_071210.mt_prots.fa... ok\n",
      "Checking phylogenomics/Proteins_renamed_r3/NC_082275.mt_prots.fa... ok\n",
      "Checking phylogenomics/Proteins_renamed_r3/NC_082276.mt_prots.fa... ok\n",
      "\n",
      "\u001b[1;32m**Step 1**\u001b[0m\n",
      "Generating indices.\n",
      "Building database for 'phylogenomics/Proteins_renamed_r3/NC_069201.mt_prots.fa'\t(14 sequences)\n",
      "Generating indices.\n",
      "Building database for 'phylogenomics/Proteins_renamed_r3/NC_082276.mt_prots.fa'\t(15 sequences)\n",
      "Generating indices.\n",
      "Building database for 'phylogenomics/Proteins_renamed_r3/NC_023128.mt_prots.fa'\t(38 sequences)\n",
      "Generating indices.\n",
      "Building database for 'phylogenomics/Proteins_renamed_r3/NC_031375.mt_prots.fa'\t(16 sequences)\n",
      "Generating indices.\n",
      "Building database for 'phylogenomics/Proteins_renamed_r3/NC_048507.mt_prots.fa'\t(14 sequences)\n",
      "Generating indices.\n",
      "Building database for 'phylogenomics/Proteins_renamed_r3/NC_033907.mt_prots.fa'\t(14 sequences)\n",
      "Generating indices.\n",
      "Building database for 'phylogenomics/Proteins_renamed_r3/NC_056195.mt_prots.fa'\t(47 sequences)\n",
      "Generating indices.\n",
      "Building database for 'phylogenomics/Proteins_renamed_r3/NC_082275.mt_prots.fa'\t(15 sequences)\n",
      "Generating indices.\n",
      "Building database for 'phylogenomics/Proteins_renamed_r3/NC_056146.mt_prots.fa'\t(73 sequences)\n",
      "Generating indices.\n",
      "Building database for 'phylogenomics/Proteins_renamed_r3/NC_027422.mt_prots.fa'\t(14 sequences)\n",
      "Generating indices.\n",
      "Building database for 'phylogenomics/Proteins_renamed_r3/NC_023127.mt_prots.fa'\t(15 sequences)\n",
      "Generating indices.\n",
      "Building database for 'phylogenomics/Proteins_renamed_r3/NC_071210.mt_prots.fa'\t(14 sequences)\n",
      "Generating indices.\n",
      "Building database for 'phylogenomics/Proteins_renamed_r3/NC_015789.mt_prots.fa'\t(15 sequences)\n",
      "Generating indices.\n",
      "Building database for 'phylogenomics/Proteins_renamed_r3/NC_070174.mt_prots.fa'\t(14 sequences)\n",
      "Generating indices.\n",
      "Building database for 'phylogenomics/Proteins_renamed_r3/NC_023125.mt_prots.fa'\t(38 sequences)\n",
      "Generating indices.\n",
      "Building database for 'phylogenomics/Proteins_renamed_r3/NC_023126.mt_prots.fa'\t(37 sequences)\n",
      "Generating indices.\n",
      "Building database for 'phylogenomics/Proteins_renamed_r3/NC_025200.mt_prots.fa'\t(15 sequences)\n",
      "Generating indices.\n",
      "Building database for 'phylogenomics/Proteins_renamed_r3/NC_056147.mt_prots.fa'\t(80 sequences)\n",
      "Generating indices.\n",
      "Building database for 'phylogenomics/Proteins_renamed_r3/NC_056148.mt_prots.fa'\t(29 sequences)\n",
      "Generating indices.\n",
      "Building database for 'phylogenomics/Proteins_renamed_r3/NC_061762.mt_prots.fa'\t(15 sequences)\n",
      "Generating indices.\n",
      "Building database for 'phylogenomics/Proteins_renamed_r3/NC_051483.mt_prots.fa'\t(64 sequences)\n",
      "Generating indices.\n",
      "Building database for 'phylogenomics/Proteins_renamed_r3/NC_015991.mt_prots.fa'\t(11 sequences)\n",
      "\n",
      "\u001b[1;32m**Step 2**\u001b[0m using diamond \n",
      "Running blast analysis: 100% (231/231)                                         \n",
      "[OUTPUT] -> written to myproject.blast-graph\n",
      "\n",
      "\u001b[1;32m**Step 3**\u001b[0m\n",
      "Clustering by similarity (Proteinortho mode) using up to 16384 MB of memory (default value, command 'free' not found) and 10 cpu core(s). Please wait...\n",
      "Clustering using 10 threads\n",
      "Reading myproject.blast-graph\n",
      "22 species\n",
      "539 paired proteins\n",
      "3368 bidirectional edges\n",
      "Clustering: working on (5@lapack* + 0@power) connected components           \n",
      "[OUTPUT] -> Orthologous groups are written to myproject.proteinortho.tsv\n",
      "[OUTPUT] -> Orthologous pairs are written to myproject.proteinortho-graph\n",
      "# myproject.proteinortho.tsv statistics:\n",
      "# number of groups that contains at least p% species:\n",
      "# p\tspecies\tgroups\tgenes in groups\n",
      "0%\t0\t80 (100%)\t539 (100%)\t\n",
      "10%\t2\t80 (100%)\t539 (100%)\t\n",
      "20%\t4\t32 (40%)\t417 (77.36%)\t\n",
      "30%\t6\t25 (31.25%)\t385 (71.42%)\t\n",
      "40%\t8\t17 (21.25%)\t324 (60.11%)\t\n",
      "50%\t11\t15 (18.75%)\t308 (57.14%)\t\n",
      "60%\t13\t14 (17.5%)\t297 (55.1%)\t\n",
      "70%\t15\t14 (17.5%)\t297 (55.1%)\t\n",
      "80%\t17\t14 (17.5%)\t297 (55.1%)\t\n",
      "90%\t19\t14 (17.5%)\t297 (55.1%)\t\n",
      "100%\t22\t6 (7.5%)\t132 (24.48%)\t\n",
      "# NOTE: p=100% correspond to the core proteome and p=0% to all groups\n",
      "# number of groups for each input species:\n",
      "NC_056147.mt_prots.fa\t56\t70%\n",
      "NC_056146.mt_prots.fa\t50\t62.5%\n",
      "NC_051483.mt_prots.fa\t42\t52.5%\n",
      "NC_056195.mt_prots.fa\t40\t50%\n",
      "NC_023128.mt_prots.fa\t37\t46.25%\n",
      "...\n",
      "NC_048507.mt_prots.fa\t14\t17.5%\n",
      "NC_069201.mt_prots.fa\t14\t17.5%\n",
      "NC_070174.mt_prots.fa\t14\t17.5%\n",
      "NC_071210.mt_prots.fa\t14\t17.5%\n",
      "NC_015991.mt_prots.fa\t10\t12.5%\n",
      "\n",
      "[OUTPUT] -> Summary is written to myproject.proteinortho-graph.summary\n",
      "[OUTPUT] -> You can extract a html version of the output using:\n",
      "proteinortho2html.pl myproject.proteinortho.tsv [PLACE FASTA FILES HERE] >myproject.proteinortho.html\n",
      "\n",
      "[OUTPUT] -> You can extract an OrthoXML (v0.3) version of the orthology group output using:\n",
      "proteinortho2xml.pl myproject.proteinortho.tsv >myproject.proteinortho.tsv.xml\n",
      "\n",
      "\u001b[1;32mAll finished.\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "! proteinortho phylogenomics/Proteins_renamed_r3/*.fa -cpus=10"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`Proteinortho` does not allow to redirect outputs, so we do it manually afterwards!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a directory to store `Proteinortho` outputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "! mkdir phylogenomics/protein_ortho_output/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Move `Proteinortho` outputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "! mv myproject* phylogenomics/protein_ortho_output/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let us work with `Proteinortho` output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the `phylogenomics/protein_ortho_output/myproject.proteinortho.tsv` file to `pandas` dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all = pd.read_csv('phylogenomics/protein_ortho_output/myproject.proteinortho.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is how it looks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># Species</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Alg.-Conn.</th>\n",
       "      <th>NC_015789.mt_prots.fa</th>\n",
       "      <th>NC_015991.mt_prots.fa</th>\n",
       "      <th>NC_023125.mt_prots.fa</th>\n",
       "      <th>NC_023126.mt_prots.fa</th>\n",
       "      <th>NC_023127.mt_prots.fa</th>\n",
       "      <th>NC_023128.mt_prots.fa</th>\n",
       "      <th>NC_025200.mt_prots.fa</th>\n",
       "      <th>...</th>\n",
       "      <th>NC_056146.mt_prots.fa</th>\n",
       "      <th>NC_056147.mt_prots.fa</th>\n",
       "      <th>NC_056148.mt_prots.fa</th>\n",
       "      <th>NC_056195.mt_prots.fa</th>\n",
       "      <th>NC_061762.mt_prots.fa</th>\n",
       "      <th>NC_069201.mt_prots.fa</th>\n",
       "      <th>NC_070174.mt_prots.fa</th>\n",
       "      <th>NC_071210.mt_prots.fa</th>\n",
       "      <th>NC_082275.mt_prots.fa</th>\n",
       "      <th>NC_082276.mt_prots.fa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NC_015789_ATP6_1</td>\n",
       "      <td>NC_015991_ATP6_1</td>\n",
       "      <td>NC_023125_ATP6_1</td>\n",
       "      <td>NC_023126_ATP6_1</td>\n",
       "      <td>NC_023127_ATP6_1</td>\n",
       "      <td>NC_023128_ATP6_1</td>\n",
       "      <td>NC_025200_ATP6_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_ATP6_1</td>\n",
       "      <td>NC_056147_ATP6_1</td>\n",
       "      <td>NC_056148_ATP6_1</td>\n",
       "      <td>NC_056195_ATP6_1</td>\n",
       "      <td>NC_061762_ATP6_1</td>\n",
       "      <td>NC_069201_ATP6_1</td>\n",
       "      <td>NC_070174_ATP6_1</td>\n",
       "      <td>NC_071210_ATP6_1</td>\n",
       "      <td>NC_082275_ATP6_1</td>\n",
       "      <td>NC_082276_ATP6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NC_015789_COX2_1</td>\n",
       "      <td>NC_015991_COX2_1</td>\n",
       "      <td>NC_023125_COX2_1</td>\n",
       "      <td>NC_023126_COX2_1</td>\n",
       "      <td>NC_023127_COX2_1</td>\n",
       "      <td>NC_023128_COX2_1</td>\n",
       "      <td>NC_025200_COX2_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_COX2_1</td>\n",
       "      <td>NC_056147_COX2_1</td>\n",
       "      <td>NC_056148_COX2_1</td>\n",
       "      <td>NC_056195_COX2_1</td>\n",
       "      <td>NC_061762_COX2_1</td>\n",
       "      <td>NC_069201_COX2_1</td>\n",
       "      <td>NC_070174_COX2_1</td>\n",
       "      <td>NC_071210_COX2_1</td>\n",
       "      <td>NC_082275_COX2_1</td>\n",
       "      <td>NC_082276_COX2_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NC_015789_COX3_1</td>\n",
       "      <td>NC_015991_COX3_1</td>\n",
       "      <td>NC_023125_COX3_1</td>\n",
       "      <td>NC_023126_COX3_1</td>\n",
       "      <td>NC_023127_COX3_1</td>\n",
       "      <td>NC_023128_COX3_1</td>\n",
       "      <td>NC_025200_COX3_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_COX3_1</td>\n",
       "      <td>NC_056147_COX3_1</td>\n",
       "      <td>NC_056148_COX3_1</td>\n",
       "      <td>NC_056195_COX3_1</td>\n",
       "      <td>NC_061762_COX3_1</td>\n",
       "      <td>NC_069201_COX3_1</td>\n",
       "      <td>NC_070174_COX3_1</td>\n",
       "      <td>NC_071210_COX3_1</td>\n",
       "      <td>NC_082275_COX3_1</td>\n",
       "      <td>NC_082276_COX3_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NC_015789_CYTB_1</td>\n",
       "      <td>NC_015991_CYTB_1</td>\n",
       "      <td>NC_023125_COB_1</td>\n",
       "      <td>NC_023126_COB_1</td>\n",
       "      <td>NC_023127_COB_1</td>\n",
       "      <td>NC_023128_COB_1</td>\n",
       "      <td>NC_025200_COB_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_COB_1</td>\n",
       "      <td>NC_056147_COB_1</td>\n",
       "      <td>NC_056148_COB_1</td>\n",
       "      <td>NC_056195_COB_1</td>\n",
       "      <td>NC_061762_COB_1</td>\n",
       "      <td>NC_069201_COB_1</td>\n",
       "      <td>NC_070174_COB_1</td>\n",
       "      <td>NC_071210_COB_1</td>\n",
       "      <td>NC_082275_COB_1</td>\n",
       "      <td>NC_082276_COB_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>0.9090</td>\n",
       "      <td>NC_015789_NAD1_1</td>\n",
       "      <td>NC_015991_NAD1_1</td>\n",
       "      <td>NC_023125_NAD1_1</td>\n",
       "      <td>NC_023126_NAD1_1</td>\n",
       "      <td>NC_023127_NAD1_1</td>\n",
       "      <td>NC_023128_NAD1_1</td>\n",
       "      <td>NC_025200_NAD1_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_NAD1_1</td>\n",
       "      <td>NC_056147_NAD1_1</td>\n",
       "      <td>NC_056148_NAD1_1</td>\n",
       "      <td>NC_056195_NAD1_1</td>\n",
       "      <td>NC_061762_NAD1_1</td>\n",
       "      <td>NC_069201_NAD1_1</td>\n",
       "      <td>NC_070174_NAD1_1</td>\n",
       "      <td>NC_071210_NAD1_1</td>\n",
       "      <td>NC_082275_NAD1_1</td>\n",
       "      <td>NC_082276_NAD1_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>0.9090</td>\n",
       "      <td>NC_015789_NAD6_1</td>\n",
       "      <td>NC_015991_NAD6_1</td>\n",
       "      <td>NC_023125_NAD6_1</td>\n",
       "      <td>NC_023126_NAD6_1</td>\n",
       "      <td>NC_023127_NAD6_1</td>\n",
       "      <td>NC_023128_NAD6_1</td>\n",
       "      <td>NC_025200_NAD6_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_NAD6_1</td>\n",
       "      <td>NC_056147_NAD6_1</td>\n",
       "      <td>NC_056148_NAD6_1</td>\n",
       "      <td>NC_056195_NAD6_1</td>\n",
       "      <td>NC_061762_NAD6_1</td>\n",
       "      <td>NC_069201_NAD6_1</td>\n",
       "      <td>NC_070174_NAD6_1</td>\n",
       "      <td>NC_071210_NAD6_1</td>\n",
       "      <td>NC_082275_NAD6_1</td>\n",
       "      <td>NC_082276_NAD6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NC_015789_ATP8_1</td>\n",
       "      <td>*</td>\n",
       "      <td>NC_023125_ATP8_1</td>\n",
       "      <td>NC_023126_ATP8_1</td>\n",
       "      <td>NC_023127_ATP8_1</td>\n",
       "      <td>NC_023128_ATP8_1</td>\n",
       "      <td>NC_025200_ATP8_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_ATP8_1</td>\n",
       "      <td>NC_056147_ATP8_1</td>\n",
       "      <td>NC_056148_ATP8_1</td>\n",
       "      <td>NC_056195_ATP8_1</td>\n",
       "      <td>NC_061762_ATP8_1</td>\n",
       "      <td>NC_069201_ATP8_1</td>\n",
       "      <td>NC_070174_ATP8_1</td>\n",
       "      <td>NC_071210_ATP8_1</td>\n",
       "      <td>NC_082275_ATP8_1</td>\n",
       "      <td>NC_082276_ATP8_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NC_015789_NAD2_1</td>\n",
       "      <td>*</td>\n",
       "      <td>NC_023125_NAD2_1</td>\n",
       "      <td>NC_023126_NAD2_1</td>\n",
       "      <td>NC_023127_NAD2_1</td>\n",
       "      <td>NC_023128_NAD2_1</td>\n",
       "      <td>NC_025200_NAD2_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_NAD2_1</td>\n",
       "      <td>NC_056147_NAD2_1</td>\n",
       "      <td>NC_056148_NAD2_1</td>\n",
       "      <td>NC_056195_NAD2_1</td>\n",
       "      <td>NC_061762_NAD2_1</td>\n",
       "      <td>NC_069201_NAD2_1</td>\n",
       "      <td>NC_070174_NAD2_1</td>\n",
       "      <td>NC_071210_NAD2_1</td>\n",
       "      <td>NC_082275_NAD2_1</td>\n",
       "      <td>NC_082276_NAD2_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NC_015789_NAD4_1</td>\n",
       "      <td>*</td>\n",
       "      <td>NC_023125_NAD4_1</td>\n",
       "      <td>NC_023126_NAD4_1</td>\n",
       "      <td>NC_023127_NAD4_1</td>\n",
       "      <td>NC_023128_NAD4_1</td>\n",
       "      <td>NC_025200_NAD4_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_NAD4_4</td>\n",
       "      <td>NC_056147_NAD4_4</td>\n",
       "      <td>NC_056148_NAD4_2</td>\n",
       "      <td>NC_056195_NAD4_1</td>\n",
       "      <td>NC_061762_NAD4_1</td>\n",
       "      <td>NC_069201_NAD4_1</td>\n",
       "      <td>NC_070174_NAD4_1</td>\n",
       "      <td>NC_071210_NAD4_2</td>\n",
       "      <td>NC_082275_NAD4_1</td>\n",
       "      <td>NC_082276_NAD4_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NC_015789_NAD4_2</td>\n",
       "      <td>NC_015991_NAD4_1</td>\n",
       "      <td>NC_023125_NAD4_2</td>\n",
       "      <td>NC_023126_NAD4_2</td>\n",
       "      <td>NC_023127_NAD4_2</td>\n",
       "      <td>NC_023128_NAD4_2</td>\n",
       "      <td>NC_025200_NAD4_2</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_NAD4_1</td>\n",
       "      <td>NC_056147_NAD4_1</td>\n",
       "      <td>NC_056148_NAD4_1</td>\n",
       "      <td>*</td>\n",
       "      <td>NC_061762_NAD4_2</td>\n",
       "      <td>NC_069201_NAD4_2</td>\n",
       "      <td>NC_070174_NAD4_2</td>\n",
       "      <td>NC_071210_NAD4_1</td>\n",
       "      <td>NC_082275_NAD4_2</td>\n",
       "      <td>NC_082276_NAD4_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>NC_015789_NAD5_1</td>\n",
       "      <td>*</td>\n",
       "      <td>NC_023125_NAD5_1</td>\n",
       "      <td>NC_023126_NAD5_1</td>\n",
       "      <td>NC_023127_NAD5_1</td>\n",
       "      <td>NC_023128_NAD5_1</td>\n",
       "      <td>NC_025200_NAD5_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_NAD5_1</td>\n",
       "      <td>NC_056147_NAD5_1</td>\n",
       "      <td>NC_056148_NAD5_1</td>\n",
       "      <td>NC_056195_NAD5_1</td>\n",
       "      <td>NC_061762_NAD5_1</td>\n",
       "      <td>NC_069201_NAD5_1</td>\n",
       "      <td>NC_070174_NAD5_1</td>\n",
       "      <td>NC_071210_NAD5_1</td>\n",
       "      <td>NC_082275_NAD5_1</td>\n",
       "      <td>NC_082276_NAD5_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>0.8570</td>\n",
       "      <td>NC_015789_NAD3_1</td>\n",
       "      <td>NC_015991_NAD3_1</td>\n",
       "      <td>NC_023125_NAD3_1</td>\n",
       "      <td>NC_023126_NAD3_1</td>\n",
       "      <td>NC_023127_NAD3_1</td>\n",
       "      <td>NC_023128_NAD3_1</td>\n",
       "      <td>NC_025200_NAD3_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_NAD3_1</td>\n",
       "      <td>NC_056147_NAD3_1</td>\n",
       "      <td>NC_056148_NAD3_1</td>\n",
       "      <td>NC_056195_NAD3_1</td>\n",
       "      <td>NC_061762_NAD3_1</td>\n",
       "      <td>NC_069201_NAD3_1</td>\n",
       "      <td>NC_070174_NAD3_1</td>\n",
       "      <td>NC_071210_NAD3_1</td>\n",
       "      <td>NC_082275_NAD3_1</td>\n",
       "      <td>NC_082276_NAD3_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>0.4980</td>\n",
       "      <td>NC_015789_COX1_1</td>\n",
       "      <td>*</td>\n",
       "      <td>NC_023125_COX1_1</td>\n",
       "      <td>NC_023126_COX1_1</td>\n",
       "      <td>NC_023127_COX1_1</td>\n",
       "      <td>NC_023128_COX1_1</td>\n",
       "      <td>NC_025200_COX1_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_COX1_1</td>\n",
       "      <td>NC_056147_COX1_1</td>\n",
       "      <td>NC_056148_COX1_1</td>\n",
       "      <td>NC_056195_COX1_1</td>\n",
       "      <td>NC_061762_COX1_1</td>\n",
       "      <td>NC_069201_COX1_1</td>\n",
       "      <td>NC_070174_COX1_1</td>\n",
       "      <td>NC_071210_COX1_1</td>\n",
       "      <td>NC_082275_COX1_1</td>\n",
       "      <td>NC_082276_COX1_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>0.0879</td>\n",
       "      <td>NC_015789_rps3_1</td>\n",
       "      <td>*</td>\n",
       "      <td>NC_023125_rps3_1</td>\n",
       "      <td>*</td>\n",
       "      <td>NC_023127_rps3_1</td>\n",
       "      <td>NC_023128_rps3_1</td>\n",
       "      <td>NC_025200_rps3_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_rps3_1</td>\n",
       "      <td>NC_056147_rps3_1</td>\n",
       "      <td>NC_056148_rps3_1</td>\n",
       "      <td>NC_056195_rps3_1</td>\n",
       "      <td>NC_061762_rps3_1</td>\n",
       "      <td>NC_069201_rps3_1</td>\n",
       "      <td>*</td>\n",
       "      <td>NC_071210_rps3_1</td>\n",
       "      <td>NC_082275_rps3_1</td>\n",
       "      <td>NC_082276_rps3_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>0.8180</td>\n",
       "      <td>NC_015789_ATP9_1</td>\n",
       "      <td>NC_015991_ATP9_1</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>NC_025200_ATP9_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_ATP9_1</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>NC_056195_ATP9_1</td>\n",
       "      <td>NC_061762_ATP9_1</td>\n",
       "      <td>*</td>\n",
       "      <td>NC_070174_ATP9_1</td>\n",
       "      <td>*</td>\n",
       "      <td>NC_082275_ATP9_1</td>\n",
       "      <td>NC_082276_ATP9_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    # Species  Genes  Alg.-Conn. NC_015789.mt_prots.fa NC_015991.mt_prots.fa  \\\n",
       "0          22     22      1.0000      NC_015789_ATP6_1      NC_015991_ATP6_1   \n",
       "1          22     22      1.0000      NC_015789_COX2_1      NC_015991_COX2_1   \n",
       "2          22     22      1.0000      NC_015789_COX3_1      NC_015991_COX3_1   \n",
       "3          22     22      1.0000      NC_015789_CYTB_1      NC_015991_CYTB_1   \n",
       "4          22     22      0.9090      NC_015789_NAD1_1      NC_015991_NAD1_1   \n",
       "5          22     22      0.9090      NC_015789_NAD6_1      NC_015991_NAD6_1   \n",
       "6          21     21      1.0000      NC_015789_ATP8_1                     *   \n",
       "7          21     21      1.0000      NC_015789_NAD2_1                     *   \n",
       "8          21     21      1.0000      NC_015789_NAD4_1                     *   \n",
       "9          21     21      1.0000      NC_015789_NAD4_2      NC_015991_NAD4_1   \n",
       "10         21     21      1.0000      NC_015789_NAD5_1                     *   \n",
       "11         21     21      0.8570      NC_015789_NAD3_1      NC_015991_NAD3_1   \n",
       "12         20     20      0.4980      NC_015789_COX1_1                     *   \n",
       "13         19     19      0.0879      NC_015789_rps3_1                     *   \n",
       "14         11     11      0.8180      NC_015789_ATP9_1      NC_015991_ATP9_1   \n",
       "\n",
       "   NC_023125.mt_prots.fa NC_023126.mt_prots.fa NC_023127.mt_prots.fa  \\\n",
       "0       NC_023125_ATP6_1      NC_023126_ATP6_1      NC_023127_ATP6_1   \n",
       "1       NC_023125_COX2_1      NC_023126_COX2_1      NC_023127_COX2_1   \n",
       "2       NC_023125_COX3_1      NC_023126_COX3_1      NC_023127_COX3_1   \n",
       "3        NC_023125_COB_1       NC_023126_COB_1       NC_023127_COB_1   \n",
       "4       NC_023125_NAD1_1      NC_023126_NAD1_1      NC_023127_NAD1_1   \n",
       "5       NC_023125_NAD6_1      NC_023126_NAD6_1      NC_023127_NAD6_1   \n",
       "6       NC_023125_ATP8_1      NC_023126_ATP8_1      NC_023127_ATP8_1   \n",
       "7       NC_023125_NAD2_1      NC_023126_NAD2_1      NC_023127_NAD2_1   \n",
       "8       NC_023125_NAD4_1      NC_023126_NAD4_1      NC_023127_NAD4_1   \n",
       "9       NC_023125_NAD4_2      NC_023126_NAD4_2      NC_023127_NAD4_2   \n",
       "10      NC_023125_NAD5_1      NC_023126_NAD5_1      NC_023127_NAD5_1   \n",
       "11      NC_023125_NAD3_1      NC_023126_NAD3_1      NC_023127_NAD3_1   \n",
       "12      NC_023125_COX1_1      NC_023126_COX1_1      NC_023127_COX1_1   \n",
       "13      NC_023125_rps3_1                     *      NC_023127_rps3_1   \n",
       "14                     *                     *                     *   \n",
       "\n",
       "   NC_023128.mt_prots.fa NC_025200.mt_prots.fa  ... NC_056146.mt_prots.fa  \\\n",
       "0       NC_023128_ATP6_1      NC_025200_ATP6_1  ...      NC_056146_ATP6_1   \n",
       "1       NC_023128_COX2_1      NC_025200_COX2_1  ...      NC_056146_COX2_1   \n",
       "2       NC_023128_COX3_1      NC_025200_COX3_1  ...      NC_056146_COX3_1   \n",
       "3        NC_023128_COB_1       NC_025200_COB_1  ...       NC_056146_COB_1   \n",
       "4       NC_023128_NAD1_1      NC_025200_NAD1_1  ...      NC_056146_NAD1_1   \n",
       "5       NC_023128_NAD6_1      NC_025200_NAD6_1  ...      NC_056146_NAD6_1   \n",
       "6       NC_023128_ATP8_1      NC_025200_ATP8_1  ...      NC_056146_ATP8_1   \n",
       "7       NC_023128_NAD2_1      NC_025200_NAD2_1  ...      NC_056146_NAD2_1   \n",
       "8       NC_023128_NAD4_1      NC_025200_NAD4_1  ...      NC_056146_NAD4_4   \n",
       "9       NC_023128_NAD4_2      NC_025200_NAD4_2  ...      NC_056146_NAD4_1   \n",
       "10      NC_023128_NAD5_1      NC_025200_NAD5_1  ...      NC_056146_NAD5_1   \n",
       "11      NC_023128_NAD3_1      NC_025200_NAD3_1  ...      NC_056146_NAD3_1   \n",
       "12      NC_023128_COX1_1      NC_025200_COX1_1  ...      NC_056146_COX1_1   \n",
       "13      NC_023128_rps3_1      NC_025200_rps3_1  ...      NC_056146_rps3_1   \n",
       "14                     *      NC_025200_ATP9_1  ...      NC_056146_ATP9_1   \n",
       "\n",
       "   NC_056147.mt_prots.fa NC_056148.mt_prots.fa NC_056195.mt_prots.fa  \\\n",
       "0       NC_056147_ATP6_1      NC_056148_ATP6_1      NC_056195_ATP6_1   \n",
       "1       NC_056147_COX2_1      NC_056148_COX2_1      NC_056195_COX2_1   \n",
       "2       NC_056147_COX3_1      NC_056148_COX3_1      NC_056195_COX3_1   \n",
       "3        NC_056147_COB_1       NC_056148_COB_1       NC_056195_COB_1   \n",
       "4       NC_056147_NAD1_1      NC_056148_NAD1_1      NC_056195_NAD1_1   \n",
       "5       NC_056147_NAD6_1      NC_056148_NAD6_1      NC_056195_NAD6_1   \n",
       "6       NC_056147_ATP8_1      NC_056148_ATP8_1      NC_056195_ATP8_1   \n",
       "7       NC_056147_NAD2_1      NC_056148_NAD2_1      NC_056195_NAD2_1   \n",
       "8       NC_056147_NAD4_4      NC_056148_NAD4_2      NC_056195_NAD4_1   \n",
       "9       NC_056147_NAD4_1      NC_056148_NAD4_1                     *   \n",
       "10      NC_056147_NAD5_1      NC_056148_NAD5_1      NC_056195_NAD5_1   \n",
       "11      NC_056147_NAD3_1      NC_056148_NAD3_1      NC_056195_NAD3_1   \n",
       "12      NC_056147_COX1_1      NC_056148_COX1_1      NC_056195_COX1_1   \n",
       "13      NC_056147_rps3_1      NC_056148_rps3_1      NC_056195_rps3_1   \n",
       "14                     *                     *      NC_056195_ATP9_1   \n",
       "\n",
       "   NC_061762.mt_prots.fa NC_069201.mt_prots.fa NC_070174.mt_prots.fa  \\\n",
       "0       NC_061762_ATP6_1      NC_069201_ATP6_1      NC_070174_ATP6_1   \n",
       "1       NC_061762_COX2_1      NC_069201_COX2_1      NC_070174_COX2_1   \n",
       "2       NC_061762_COX3_1      NC_069201_COX3_1      NC_070174_COX3_1   \n",
       "3        NC_061762_COB_1       NC_069201_COB_1       NC_070174_COB_1   \n",
       "4       NC_061762_NAD1_1      NC_069201_NAD1_1      NC_070174_NAD1_1   \n",
       "5       NC_061762_NAD6_1      NC_069201_NAD6_1      NC_070174_NAD6_1   \n",
       "6       NC_061762_ATP8_1      NC_069201_ATP8_1      NC_070174_ATP8_1   \n",
       "7       NC_061762_NAD2_1      NC_069201_NAD2_1      NC_070174_NAD2_1   \n",
       "8       NC_061762_NAD4_1      NC_069201_NAD4_1      NC_070174_NAD4_1   \n",
       "9       NC_061762_NAD4_2      NC_069201_NAD4_2      NC_070174_NAD4_2   \n",
       "10      NC_061762_NAD5_1      NC_069201_NAD5_1      NC_070174_NAD5_1   \n",
       "11      NC_061762_NAD3_1      NC_069201_NAD3_1      NC_070174_NAD3_1   \n",
       "12      NC_061762_COX1_1      NC_069201_COX1_1      NC_070174_COX1_1   \n",
       "13      NC_061762_rps3_1      NC_069201_rps3_1                     *   \n",
       "14      NC_061762_ATP9_1                     *      NC_070174_ATP9_1   \n",
       "\n",
       "   NC_071210.mt_prots.fa NC_082275.mt_prots.fa NC_082276.mt_prots.fa  \n",
       "0       NC_071210_ATP6_1      NC_082275_ATP6_1      NC_082276_ATP6_1  \n",
       "1       NC_071210_COX2_1      NC_082275_COX2_1      NC_082276_COX2_1  \n",
       "2       NC_071210_COX3_1      NC_082275_COX3_1      NC_082276_COX3_1  \n",
       "3        NC_071210_COB_1       NC_082275_COB_1       NC_082276_COB_1  \n",
       "4       NC_071210_NAD1_1      NC_082275_NAD1_1      NC_082276_NAD1_1  \n",
       "5       NC_071210_NAD6_1      NC_082275_NAD6_1      NC_082276_NAD6_1  \n",
       "6       NC_071210_ATP8_1      NC_082275_ATP8_1      NC_082276_ATP8_1  \n",
       "7       NC_071210_NAD2_1      NC_082275_NAD2_1      NC_082276_NAD2_1  \n",
       "8       NC_071210_NAD4_2      NC_082275_NAD4_1      NC_082276_NAD4_1  \n",
       "9       NC_071210_NAD4_1      NC_082275_NAD4_2      NC_082276_NAD4_2  \n",
       "10      NC_071210_NAD5_1      NC_082275_NAD5_1      NC_082276_NAD5_1  \n",
       "11      NC_071210_NAD3_1      NC_082275_NAD3_1      NC_082276_NAD3_1  \n",
       "12      NC_071210_COX1_1      NC_082275_COX1_1      NC_082276_COX1_1  \n",
       "13      NC_071210_rps3_1      NC_082275_rps3_1      NC_082276_rps3_1  \n",
       "14                     *      NC_082275_ATP9_1      NC_082276_ATP9_1  \n",
       "\n",
       "[15 rows x 25 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all.head(15)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We shall keep only the first 14 rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all = df_all.iloc[:13]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># Species</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Alg.-Conn.</th>\n",
       "      <th>NC_015789.mt_prots.fa</th>\n",
       "      <th>NC_015991.mt_prots.fa</th>\n",
       "      <th>NC_023125.mt_prots.fa</th>\n",
       "      <th>NC_023126.mt_prots.fa</th>\n",
       "      <th>NC_023127.mt_prots.fa</th>\n",
       "      <th>NC_023128.mt_prots.fa</th>\n",
       "      <th>NC_025200.mt_prots.fa</th>\n",
       "      <th>...</th>\n",
       "      <th>NC_056146.mt_prots.fa</th>\n",
       "      <th>NC_056147.mt_prots.fa</th>\n",
       "      <th>NC_056148.mt_prots.fa</th>\n",
       "      <th>NC_056195.mt_prots.fa</th>\n",
       "      <th>NC_061762.mt_prots.fa</th>\n",
       "      <th>NC_069201.mt_prots.fa</th>\n",
       "      <th>NC_070174.mt_prots.fa</th>\n",
       "      <th>NC_071210.mt_prots.fa</th>\n",
       "      <th>NC_082275.mt_prots.fa</th>\n",
       "      <th>NC_082276.mt_prots.fa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>1.000</td>\n",
       "      <td>NC_015789_ATP6_1</td>\n",
       "      <td>NC_015991_ATP6_1</td>\n",
       "      <td>NC_023125_ATP6_1</td>\n",
       "      <td>NC_023126_ATP6_1</td>\n",
       "      <td>NC_023127_ATP6_1</td>\n",
       "      <td>NC_023128_ATP6_1</td>\n",
       "      <td>NC_025200_ATP6_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_ATP6_1</td>\n",
       "      <td>NC_056147_ATP6_1</td>\n",
       "      <td>NC_056148_ATP6_1</td>\n",
       "      <td>NC_056195_ATP6_1</td>\n",
       "      <td>NC_061762_ATP6_1</td>\n",
       "      <td>NC_069201_ATP6_1</td>\n",
       "      <td>NC_070174_ATP6_1</td>\n",
       "      <td>NC_071210_ATP6_1</td>\n",
       "      <td>NC_082275_ATP6_1</td>\n",
       "      <td>NC_082276_ATP6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>1.000</td>\n",
       "      <td>NC_015789_COX2_1</td>\n",
       "      <td>NC_015991_COX2_1</td>\n",
       "      <td>NC_023125_COX2_1</td>\n",
       "      <td>NC_023126_COX2_1</td>\n",
       "      <td>NC_023127_COX2_1</td>\n",
       "      <td>NC_023128_COX2_1</td>\n",
       "      <td>NC_025200_COX2_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_COX2_1</td>\n",
       "      <td>NC_056147_COX2_1</td>\n",
       "      <td>NC_056148_COX2_1</td>\n",
       "      <td>NC_056195_COX2_1</td>\n",
       "      <td>NC_061762_COX2_1</td>\n",
       "      <td>NC_069201_COX2_1</td>\n",
       "      <td>NC_070174_COX2_1</td>\n",
       "      <td>NC_071210_COX2_1</td>\n",
       "      <td>NC_082275_COX2_1</td>\n",
       "      <td>NC_082276_COX2_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>1.000</td>\n",
       "      <td>NC_015789_COX3_1</td>\n",
       "      <td>NC_015991_COX3_1</td>\n",
       "      <td>NC_023125_COX3_1</td>\n",
       "      <td>NC_023126_COX3_1</td>\n",
       "      <td>NC_023127_COX3_1</td>\n",
       "      <td>NC_023128_COX3_1</td>\n",
       "      <td>NC_025200_COX3_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_COX3_1</td>\n",
       "      <td>NC_056147_COX3_1</td>\n",
       "      <td>NC_056148_COX3_1</td>\n",
       "      <td>NC_056195_COX3_1</td>\n",
       "      <td>NC_061762_COX3_1</td>\n",
       "      <td>NC_069201_COX3_1</td>\n",
       "      <td>NC_070174_COX3_1</td>\n",
       "      <td>NC_071210_COX3_1</td>\n",
       "      <td>NC_082275_COX3_1</td>\n",
       "      <td>NC_082276_COX3_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>1.000</td>\n",
       "      <td>NC_015789_CYTB_1</td>\n",
       "      <td>NC_015991_CYTB_1</td>\n",
       "      <td>NC_023125_COB_1</td>\n",
       "      <td>NC_023126_COB_1</td>\n",
       "      <td>NC_023127_COB_1</td>\n",
       "      <td>NC_023128_COB_1</td>\n",
       "      <td>NC_025200_COB_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_COB_1</td>\n",
       "      <td>NC_056147_COB_1</td>\n",
       "      <td>NC_056148_COB_1</td>\n",
       "      <td>NC_056195_COB_1</td>\n",
       "      <td>NC_061762_COB_1</td>\n",
       "      <td>NC_069201_COB_1</td>\n",
       "      <td>NC_070174_COB_1</td>\n",
       "      <td>NC_071210_COB_1</td>\n",
       "      <td>NC_082275_COB_1</td>\n",
       "      <td>NC_082276_COB_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>0.909</td>\n",
       "      <td>NC_015789_NAD1_1</td>\n",
       "      <td>NC_015991_NAD1_1</td>\n",
       "      <td>NC_023125_NAD1_1</td>\n",
       "      <td>NC_023126_NAD1_1</td>\n",
       "      <td>NC_023127_NAD1_1</td>\n",
       "      <td>NC_023128_NAD1_1</td>\n",
       "      <td>NC_025200_NAD1_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_NAD1_1</td>\n",
       "      <td>NC_056147_NAD1_1</td>\n",
       "      <td>NC_056148_NAD1_1</td>\n",
       "      <td>NC_056195_NAD1_1</td>\n",
       "      <td>NC_061762_NAD1_1</td>\n",
       "      <td>NC_069201_NAD1_1</td>\n",
       "      <td>NC_070174_NAD1_1</td>\n",
       "      <td>NC_071210_NAD1_1</td>\n",
       "      <td>NC_082275_NAD1_1</td>\n",
       "      <td>NC_082276_NAD1_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>0.909</td>\n",
       "      <td>NC_015789_NAD6_1</td>\n",
       "      <td>NC_015991_NAD6_1</td>\n",
       "      <td>NC_023125_NAD6_1</td>\n",
       "      <td>NC_023126_NAD6_1</td>\n",
       "      <td>NC_023127_NAD6_1</td>\n",
       "      <td>NC_023128_NAD6_1</td>\n",
       "      <td>NC_025200_NAD6_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_NAD6_1</td>\n",
       "      <td>NC_056147_NAD6_1</td>\n",
       "      <td>NC_056148_NAD6_1</td>\n",
       "      <td>NC_056195_NAD6_1</td>\n",
       "      <td>NC_061762_NAD6_1</td>\n",
       "      <td>NC_069201_NAD6_1</td>\n",
       "      <td>NC_070174_NAD6_1</td>\n",
       "      <td>NC_071210_NAD6_1</td>\n",
       "      <td>NC_082275_NAD6_1</td>\n",
       "      <td>NC_082276_NAD6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>1.000</td>\n",
       "      <td>NC_015789_ATP8_1</td>\n",
       "      <td>*</td>\n",
       "      <td>NC_023125_ATP8_1</td>\n",
       "      <td>NC_023126_ATP8_1</td>\n",
       "      <td>NC_023127_ATP8_1</td>\n",
       "      <td>NC_023128_ATP8_1</td>\n",
       "      <td>NC_025200_ATP8_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_ATP8_1</td>\n",
       "      <td>NC_056147_ATP8_1</td>\n",
       "      <td>NC_056148_ATP8_1</td>\n",
       "      <td>NC_056195_ATP8_1</td>\n",
       "      <td>NC_061762_ATP8_1</td>\n",
       "      <td>NC_069201_ATP8_1</td>\n",
       "      <td>NC_070174_ATP8_1</td>\n",
       "      <td>NC_071210_ATP8_1</td>\n",
       "      <td>NC_082275_ATP8_1</td>\n",
       "      <td>NC_082276_ATP8_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>1.000</td>\n",
       "      <td>NC_015789_NAD2_1</td>\n",
       "      <td>*</td>\n",
       "      <td>NC_023125_NAD2_1</td>\n",
       "      <td>NC_023126_NAD2_1</td>\n",
       "      <td>NC_023127_NAD2_1</td>\n",
       "      <td>NC_023128_NAD2_1</td>\n",
       "      <td>NC_025200_NAD2_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_NAD2_1</td>\n",
       "      <td>NC_056147_NAD2_1</td>\n",
       "      <td>NC_056148_NAD2_1</td>\n",
       "      <td>NC_056195_NAD2_1</td>\n",
       "      <td>NC_061762_NAD2_1</td>\n",
       "      <td>NC_069201_NAD2_1</td>\n",
       "      <td>NC_070174_NAD2_1</td>\n",
       "      <td>NC_071210_NAD2_1</td>\n",
       "      <td>NC_082275_NAD2_1</td>\n",
       "      <td>NC_082276_NAD2_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>1.000</td>\n",
       "      <td>NC_015789_NAD4_1</td>\n",
       "      <td>*</td>\n",
       "      <td>NC_023125_NAD4_1</td>\n",
       "      <td>NC_023126_NAD4_1</td>\n",
       "      <td>NC_023127_NAD4_1</td>\n",
       "      <td>NC_023128_NAD4_1</td>\n",
       "      <td>NC_025200_NAD4_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_NAD4_4</td>\n",
       "      <td>NC_056147_NAD4_4</td>\n",
       "      <td>NC_056148_NAD4_2</td>\n",
       "      <td>NC_056195_NAD4_1</td>\n",
       "      <td>NC_061762_NAD4_1</td>\n",
       "      <td>NC_069201_NAD4_1</td>\n",
       "      <td>NC_070174_NAD4_1</td>\n",
       "      <td>NC_071210_NAD4_2</td>\n",
       "      <td>NC_082275_NAD4_1</td>\n",
       "      <td>NC_082276_NAD4_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>1.000</td>\n",
       "      <td>NC_015789_NAD4_2</td>\n",
       "      <td>NC_015991_NAD4_1</td>\n",
       "      <td>NC_023125_NAD4_2</td>\n",
       "      <td>NC_023126_NAD4_2</td>\n",
       "      <td>NC_023127_NAD4_2</td>\n",
       "      <td>NC_023128_NAD4_2</td>\n",
       "      <td>NC_025200_NAD4_2</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_NAD4_1</td>\n",
       "      <td>NC_056147_NAD4_1</td>\n",
       "      <td>NC_056148_NAD4_1</td>\n",
       "      <td>*</td>\n",
       "      <td>NC_061762_NAD4_2</td>\n",
       "      <td>NC_069201_NAD4_2</td>\n",
       "      <td>NC_070174_NAD4_2</td>\n",
       "      <td>NC_071210_NAD4_1</td>\n",
       "      <td>NC_082275_NAD4_2</td>\n",
       "      <td>NC_082276_NAD4_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>1.000</td>\n",
       "      <td>NC_015789_NAD5_1</td>\n",
       "      <td>*</td>\n",
       "      <td>NC_023125_NAD5_1</td>\n",
       "      <td>NC_023126_NAD5_1</td>\n",
       "      <td>NC_023127_NAD5_1</td>\n",
       "      <td>NC_023128_NAD5_1</td>\n",
       "      <td>NC_025200_NAD5_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_NAD5_1</td>\n",
       "      <td>NC_056147_NAD5_1</td>\n",
       "      <td>NC_056148_NAD5_1</td>\n",
       "      <td>NC_056195_NAD5_1</td>\n",
       "      <td>NC_061762_NAD5_1</td>\n",
       "      <td>NC_069201_NAD5_1</td>\n",
       "      <td>NC_070174_NAD5_1</td>\n",
       "      <td>NC_071210_NAD5_1</td>\n",
       "      <td>NC_082275_NAD5_1</td>\n",
       "      <td>NC_082276_NAD5_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>0.857</td>\n",
       "      <td>NC_015789_NAD3_1</td>\n",
       "      <td>NC_015991_NAD3_1</td>\n",
       "      <td>NC_023125_NAD3_1</td>\n",
       "      <td>NC_023126_NAD3_1</td>\n",
       "      <td>NC_023127_NAD3_1</td>\n",
       "      <td>NC_023128_NAD3_1</td>\n",
       "      <td>NC_025200_NAD3_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_NAD3_1</td>\n",
       "      <td>NC_056147_NAD3_1</td>\n",
       "      <td>NC_056148_NAD3_1</td>\n",
       "      <td>NC_056195_NAD3_1</td>\n",
       "      <td>NC_061762_NAD3_1</td>\n",
       "      <td>NC_069201_NAD3_1</td>\n",
       "      <td>NC_070174_NAD3_1</td>\n",
       "      <td>NC_071210_NAD3_1</td>\n",
       "      <td>NC_082275_NAD3_1</td>\n",
       "      <td>NC_082276_NAD3_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>0.498</td>\n",
       "      <td>NC_015789_COX1_1</td>\n",
       "      <td>*</td>\n",
       "      <td>NC_023125_COX1_1</td>\n",
       "      <td>NC_023126_COX1_1</td>\n",
       "      <td>NC_023127_COX1_1</td>\n",
       "      <td>NC_023128_COX1_1</td>\n",
       "      <td>NC_025200_COX1_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_COX1_1</td>\n",
       "      <td>NC_056147_COX1_1</td>\n",
       "      <td>NC_056148_COX1_1</td>\n",
       "      <td>NC_056195_COX1_1</td>\n",
       "      <td>NC_061762_COX1_1</td>\n",
       "      <td>NC_069201_COX1_1</td>\n",
       "      <td>NC_070174_COX1_1</td>\n",
       "      <td>NC_071210_COX1_1</td>\n",
       "      <td>NC_082275_COX1_1</td>\n",
       "      <td>NC_082276_COX1_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    # Species  Genes  Alg.-Conn. NC_015789.mt_prots.fa NC_015991.mt_prots.fa  \\\n",
       "0          22     22       1.000      NC_015789_ATP6_1      NC_015991_ATP6_1   \n",
       "1          22     22       1.000      NC_015789_COX2_1      NC_015991_COX2_1   \n",
       "2          22     22       1.000      NC_015789_COX3_1      NC_015991_COX3_1   \n",
       "3          22     22       1.000      NC_015789_CYTB_1      NC_015991_CYTB_1   \n",
       "4          22     22       0.909      NC_015789_NAD1_1      NC_015991_NAD1_1   \n",
       "5          22     22       0.909      NC_015789_NAD6_1      NC_015991_NAD6_1   \n",
       "6          21     21       1.000      NC_015789_ATP8_1                     *   \n",
       "7          21     21       1.000      NC_015789_NAD2_1                     *   \n",
       "8          21     21       1.000      NC_015789_NAD4_1                     *   \n",
       "9          21     21       1.000      NC_015789_NAD4_2      NC_015991_NAD4_1   \n",
       "10         21     21       1.000      NC_015789_NAD5_1                     *   \n",
       "11         21     21       0.857      NC_015789_NAD3_1      NC_015991_NAD3_1   \n",
       "12         20     20       0.498      NC_015789_COX1_1                     *   \n",
       "\n",
       "   NC_023125.mt_prots.fa NC_023126.mt_prots.fa NC_023127.mt_prots.fa  \\\n",
       "0       NC_023125_ATP6_1      NC_023126_ATP6_1      NC_023127_ATP6_1   \n",
       "1       NC_023125_COX2_1      NC_023126_COX2_1      NC_023127_COX2_1   \n",
       "2       NC_023125_COX3_1      NC_023126_COX3_1      NC_023127_COX3_1   \n",
       "3        NC_023125_COB_1       NC_023126_COB_1       NC_023127_COB_1   \n",
       "4       NC_023125_NAD1_1      NC_023126_NAD1_1      NC_023127_NAD1_1   \n",
       "5       NC_023125_NAD6_1      NC_023126_NAD6_1      NC_023127_NAD6_1   \n",
       "6       NC_023125_ATP8_1      NC_023126_ATP8_1      NC_023127_ATP8_1   \n",
       "7       NC_023125_NAD2_1      NC_023126_NAD2_1      NC_023127_NAD2_1   \n",
       "8       NC_023125_NAD4_1      NC_023126_NAD4_1      NC_023127_NAD4_1   \n",
       "9       NC_023125_NAD4_2      NC_023126_NAD4_2      NC_023127_NAD4_2   \n",
       "10      NC_023125_NAD5_1      NC_023126_NAD5_1      NC_023127_NAD5_1   \n",
       "11      NC_023125_NAD3_1      NC_023126_NAD3_1      NC_023127_NAD3_1   \n",
       "12      NC_023125_COX1_1      NC_023126_COX1_1      NC_023127_COX1_1   \n",
       "\n",
       "   NC_023128.mt_prots.fa NC_025200.mt_prots.fa  ... NC_056146.mt_prots.fa  \\\n",
       "0       NC_023128_ATP6_1      NC_025200_ATP6_1  ...      NC_056146_ATP6_1   \n",
       "1       NC_023128_COX2_1      NC_025200_COX2_1  ...      NC_056146_COX2_1   \n",
       "2       NC_023128_COX3_1      NC_025200_COX3_1  ...      NC_056146_COX3_1   \n",
       "3        NC_023128_COB_1       NC_025200_COB_1  ...       NC_056146_COB_1   \n",
       "4       NC_023128_NAD1_1      NC_025200_NAD1_1  ...      NC_056146_NAD1_1   \n",
       "5       NC_023128_NAD6_1      NC_025200_NAD6_1  ...      NC_056146_NAD6_1   \n",
       "6       NC_023128_ATP8_1      NC_025200_ATP8_1  ...      NC_056146_ATP8_1   \n",
       "7       NC_023128_NAD2_1      NC_025200_NAD2_1  ...      NC_056146_NAD2_1   \n",
       "8       NC_023128_NAD4_1      NC_025200_NAD4_1  ...      NC_056146_NAD4_4   \n",
       "9       NC_023128_NAD4_2      NC_025200_NAD4_2  ...      NC_056146_NAD4_1   \n",
       "10      NC_023128_NAD5_1      NC_025200_NAD5_1  ...      NC_056146_NAD5_1   \n",
       "11      NC_023128_NAD3_1      NC_025200_NAD3_1  ...      NC_056146_NAD3_1   \n",
       "12      NC_023128_COX1_1      NC_025200_COX1_1  ...      NC_056146_COX1_1   \n",
       "\n",
       "   NC_056147.mt_prots.fa NC_056148.mt_prots.fa NC_056195.mt_prots.fa  \\\n",
       "0       NC_056147_ATP6_1      NC_056148_ATP6_1      NC_056195_ATP6_1   \n",
       "1       NC_056147_COX2_1      NC_056148_COX2_1      NC_056195_COX2_1   \n",
       "2       NC_056147_COX3_1      NC_056148_COX3_1      NC_056195_COX3_1   \n",
       "3        NC_056147_COB_1       NC_056148_COB_1       NC_056195_COB_1   \n",
       "4       NC_056147_NAD1_1      NC_056148_NAD1_1      NC_056195_NAD1_1   \n",
       "5       NC_056147_NAD6_1      NC_056148_NAD6_1      NC_056195_NAD6_1   \n",
       "6       NC_056147_ATP8_1      NC_056148_ATP8_1      NC_056195_ATP8_1   \n",
       "7       NC_056147_NAD2_1      NC_056148_NAD2_1      NC_056195_NAD2_1   \n",
       "8       NC_056147_NAD4_4      NC_056148_NAD4_2      NC_056195_NAD4_1   \n",
       "9       NC_056147_NAD4_1      NC_056148_NAD4_1                     *   \n",
       "10      NC_056147_NAD5_1      NC_056148_NAD5_1      NC_056195_NAD5_1   \n",
       "11      NC_056147_NAD3_1      NC_056148_NAD3_1      NC_056195_NAD3_1   \n",
       "12      NC_056147_COX1_1      NC_056148_COX1_1      NC_056195_COX1_1   \n",
       "\n",
       "   NC_061762.mt_prots.fa NC_069201.mt_prots.fa NC_070174.mt_prots.fa  \\\n",
       "0       NC_061762_ATP6_1      NC_069201_ATP6_1      NC_070174_ATP6_1   \n",
       "1       NC_061762_COX2_1      NC_069201_COX2_1      NC_070174_COX2_1   \n",
       "2       NC_061762_COX3_1      NC_069201_COX3_1      NC_070174_COX3_1   \n",
       "3        NC_061762_COB_1       NC_069201_COB_1       NC_070174_COB_1   \n",
       "4       NC_061762_NAD1_1      NC_069201_NAD1_1      NC_070174_NAD1_1   \n",
       "5       NC_061762_NAD6_1      NC_069201_NAD6_1      NC_070174_NAD6_1   \n",
       "6       NC_061762_ATP8_1      NC_069201_ATP8_1      NC_070174_ATP8_1   \n",
       "7       NC_061762_NAD2_1      NC_069201_NAD2_1      NC_070174_NAD2_1   \n",
       "8       NC_061762_NAD4_1      NC_069201_NAD4_1      NC_070174_NAD4_1   \n",
       "9       NC_061762_NAD4_2      NC_069201_NAD4_2      NC_070174_NAD4_2   \n",
       "10      NC_061762_NAD5_1      NC_069201_NAD5_1      NC_070174_NAD5_1   \n",
       "11      NC_061762_NAD3_1      NC_069201_NAD3_1      NC_070174_NAD3_1   \n",
       "12      NC_061762_COX1_1      NC_069201_COX1_1      NC_070174_COX1_1   \n",
       "\n",
       "   NC_071210.mt_prots.fa NC_082275.mt_prots.fa NC_082276.mt_prots.fa  \n",
       "0       NC_071210_ATP6_1      NC_082275_ATP6_1      NC_082276_ATP6_1  \n",
       "1       NC_071210_COX2_1      NC_082275_COX2_1      NC_082276_COX2_1  \n",
       "2       NC_071210_COX3_1      NC_082275_COX3_1      NC_082276_COX3_1  \n",
       "3        NC_071210_COB_1       NC_082275_COB_1       NC_082276_COB_1  \n",
       "4       NC_071210_NAD1_1      NC_082275_NAD1_1      NC_082276_NAD1_1  \n",
       "5       NC_071210_NAD6_1      NC_082275_NAD6_1      NC_082276_NAD6_1  \n",
       "6       NC_071210_ATP8_1      NC_082275_ATP8_1      NC_082276_ATP8_1  \n",
       "7       NC_071210_NAD2_1      NC_082275_NAD2_1      NC_082276_NAD2_1  \n",
       "8       NC_071210_NAD4_2      NC_082275_NAD4_1      NC_082276_NAD4_1  \n",
       "9       NC_071210_NAD4_1      NC_082275_NAD4_2      NC_082276_NAD4_2  \n",
       "10      NC_071210_NAD5_1      NC_082275_NAD5_1      NC_082276_NAD5_1  \n",
       "11      NC_071210_NAD3_1      NC_082275_NAD3_1      NC_082276_NAD3_1  \n",
       "12      NC_071210_COX1_1      NC_082275_COX1_1      NC_082276_COX1_1  \n",
       "\n",
       "[13 rows x 25 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Good."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If there are any non-single copy orthologs we must delete them!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all.loc[1] = df_all.loc[1].apply(lambda x: x.replace(',', '*') if isinstance(x, str) else x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># Species</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Alg.-Conn.</th>\n",
       "      <th>NC_015789.mt_prots.fa</th>\n",
       "      <th>NC_015991.mt_prots.fa</th>\n",
       "      <th>NC_023125.mt_prots.fa</th>\n",
       "      <th>NC_023126.mt_prots.fa</th>\n",
       "      <th>NC_023127.mt_prots.fa</th>\n",
       "      <th>NC_023128.mt_prots.fa</th>\n",
       "      <th>NC_025200.mt_prots.fa</th>\n",
       "      <th>...</th>\n",
       "      <th>NC_056146.mt_prots.fa</th>\n",
       "      <th>NC_056147.mt_prots.fa</th>\n",
       "      <th>NC_056148.mt_prots.fa</th>\n",
       "      <th>NC_056195.mt_prots.fa</th>\n",
       "      <th>NC_061762.mt_prots.fa</th>\n",
       "      <th>NC_069201.mt_prots.fa</th>\n",
       "      <th>NC_070174.mt_prots.fa</th>\n",
       "      <th>NC_071210.mt_prots.fa</th>\n",
       "      <th>NC_082275.mt_prots.fa</th>\n",
       "      <th>NC_082276.mt_prots.fa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>1.000</td>\n",
       "      <td>NC_015789_ATP6_1</td>\n",
       "      <td>NC_015991_ATP6_1</td>\n",
       "      <td>NC_023125_ATP6_1</td>\n",
       "      <td>NC_023126_ATP6_1</td>\n",
       "      <td>NC_023127_ATP6_1</td>\n",
       "      <td>NC_023128_ATP6_1</td>\n",
       "      <td>NC_025200_ATP6_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_ATP6_1</td>\n",
       "      <td>NC_056147_ATP6_1</td>\n",
       "      <td>NC_056148_ATP6_1</td>\n",
       "      <td>NC_056195_ATP6_1</td>\n",
       "      <td>NC_061762_ATP6_1</td>\n",
       "      <td>NC_069201_ATP6_1</td>\n",
       "      <td>NC_070174_ATP6_1</td>\n",
       "      <td>NC_071210_ATP6_1</td>\n",
       "      <td>NC_082275_ATP6_1</td>\n",
       "      <td>NC_082276_ATP6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>1.000</td>\n",
       "      <td>NC_015789_COX2_1</td>\n",
       "      <td>NC_015991_COX2_1</td>\n",
       "      <td>NC_023125_COX2_1</td>\n",
       "      <td>NC_023126_COX2_1</td>\n",
       "      <td>NC_023127_COX2_1</td>\n",
       "      <td>NC_023128_COX2_1</td>\n",
       "      <td>NC_025200_COX2_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_COX2_1</td>\n",
       "      <td>NC_056147_COX2_1</td>\n",
       "      <td>NC_056148_COX2_1</td>\n",
       "      <td>NC_056195_COX2_1</td>\n",
       "      <td>NC_061762_COX2_1</td>\n",
       "      <td>NC_069201_COX2_1</td>\n",
       "      <td>NC_070174_COX2_1</td>\n",
       "      <td>NC_071210_COX2_1</td>\n",
       "      <td>NC_082275_COX2_1</td>\n",
       "      <td>NC_082276_COX2_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>1.000</td>\n",
       "      <td>NC_015789_COX3_1</td>\n",
       "      <td>NC_015991_COX3_1</td>\n",
       "      <td>NC_023125_COX3_1</td>\n",
       "      <td>NC_023126_COX3_1</td>\n",
       "      <td>NC_023127_COX3_1</td>\n",
       "      <td>NC_023128_COX3_1</td>\n",
       "      <td>NC_025200_COX3_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_COX3_1</td>\n",
       "      <td>NC_056147_COX3_1</td>\n",
       "      <td>NC_056148_COX3_1</td>\n",
       "      <td>NC_056195_COX3_1</td>\n",
       "      <td>NC_061762_COX3_1</td>\n",
       "      <td>NC_069201_COX3_1</td>\n",
       "      <td>NC_070174_COX3_1</td>\n",
       "      <td>NC_071210_COX3_1</td>\n",
       "      <td>NC_082275_COX3_1</td>\n",
       "      <td>NC_082276_COX3_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>1.000</td>\n",
       "      <td>NC_015789_CYTB_1</td>\n",
       "      <td>NC_015991_CYTB_1</td>\n",
       "      <td>NC_023125_COB_1</td>\n",
       "      <td>NC_023126_COB_1</td>\n",
       "      <td>NC_023127_COB_1</td>\n",
       "      <td>NC_023128_COB_1</td>\n",
       "      <td>NC_025200_COB_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_COB_1</td>\n",
       "      <td>NC_056147_COB_1</td>\n",
       "      <td>NC_056148_COB_1</td>\n",
       "      <td>NC_056195_COB_1</td>\n",
       "      <td>NC_061762_COB_1</td>\n",
       "      <td>NC_069201_COB_1</td>\n",
       "      <td>NC_070174_COB_1</td>\n",
       "      <td>NC_071210_COB_1</td>\n",
       "      <td>NC_082275_COB_1</td>\n",
       "      <td>NC_082276_COB_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>0.909</td>\n",
       "      <td>NC_015789_NAD1_1</td>\n",
       "      <td>NC_015991_NAD1_1</td>\n",
       "      <td>NC_023125_NAD1_1</td>\n",
       "      <td>NC_023126_NAD1_1</td>\n",
       "      <td>NC_023127_NAD1_1</td>\n",
       "      <td>NC_023128_NAD1_1</td>\n",
       "      <td>NC_025200_NAD1_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_NAD1_1</td>\n",
       "      <td>NC_056147_NAD1_1</td>\n",
       "      <td>NC_056148_NAD1_1</td>\n",
       "      <td>NC_056195_NAD1_1</td>\n",
       "      <td>NC_061762_NAD1_1</td>\n",
       "      <td>NC_069201_NAD1_1</td>\n",
       "      <td>NC_070174_NAD1_1</td>\n",
       "      <td>NC_071210_NAD1_1</td>\n",
       "      <td>NC_082275_NAD1_1</td>\n",
       "      <td>NC_082276_NAD1_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>0.909</td>\n",
       "      <td>NC_015789_NAD6_1</td>\n",
       "      <td>NC_015991_NAD6_1</td>\n",
       "      <td>NC_023125_NAD6_1</td>\n",
       "      <td>NC_023126_NAD6_1</td>\n",
       "      <td>NC_023127_NAD6_1</td>\n",
       "      <td>NC_023128_NAD6_1</td>\n",
       "      <td>NC_025200_NAD6_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_NAD6_1</td>\n",
       "      <td>NC_056147_NAD6_1</td>\n",
       "      <td>NC_056148_NAD6_1</td>\n",
       "      <td>NC_056195_NAD6_1</td>\n",
       "      <td>NC_061762_NAD6_1</td>\n",
       "      <td>NC_069201_NAD6_1</td>\n",
       "      <td>NC_070174_NAD6_1</td>\n",
       "      <td>NC_071210_NAD6_1</td>\n",
       "      <td>NC_082275_NAD6_1</td>\n",
       "      <td>NC_082276_NAD6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>1.000</td>\n",
       "      <td>NC_015789_ATP8_1</td>\n",
       "      <td>*</td>\n",
       "      <td>NC_023125_ATP8_1</td>\n",
       "      <td>NC_023126_ATP8_1</td>\n",
       "      <td>NC_023127_ATP8_1</td>\n",
       "      <td>NC_023128_ATP8_1</td>\n",
       "      <td>NC_025200_ATP8_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_ATP8_1</td>\n",
       "      <td>NC_056147_ATP8_1</td>\n",
       "      <td>NC_056148_ATP8_1</td>\n",
       "      <td>NC_056195_ATP8_1</td>\n",
       "      <td>NC_061762_ATP8_1</td>\n",
       "      <td>NC_069201_ATP8_1</td>\n",
       "      <td>NC_070174_ATP8_1</td>\n",
       "      <td>NC_071210_ATP8_1</td>\n",
       "      <td>NC_082275_ATP8_1</td>\n",
       "      <td>NC_082276_ATP8_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>1.000</td>\n",
       "      <td>NC_015789_NAD2_1</td>\n",
       "      <td>*</td>\n",
       "      <td>NC_023125_NAD2_1</td>\n",
       "      <td>NC_023126_NAD2_1</td>\n",
       "      <td>NC_023127_NAD2_1</td>\n",
       "      <td>NC_023128_NAD2_1</td>\n",
       "      <td>NC_025200_NAD2_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_NAD2_1</td>\n",
       "      <td>NC_056147_NAD2_1</td>\n",
       "      <td>NC_056148_NAD2_1</td>\n",
       "      <td>NC_056195_NAD2_1</td>\n",
       "      <td>NC_061762_NAD2_1</td>\n",
       "      <td>NC_069201_NAD2_1</td>\n",
       "      <td>NC_070174_NAD2_1</td>\n",
       "      <td>NC_071210_NAD2_1</td>\n",
       "      <td>NC_082275_NAD2_1</td>\n",
       "      <td>NC_082276_NAD2_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>1.000</td>\n",
       "      <td>NC_015789_NAD4_1</td>\n",
       "      <td>*</td>\n",
       "      <td>NC_023125_NAD4_1</td>\n",
       "      <td>NC_023126_NAD4_1</td>\n",
       "      <td>NC_023127_NAD4_1</td>\n",
       "      <td>NC_023128_NAD4_1</td>\n",
       "      <td>NC_025200_NAD4_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_NAD4_4</td>\n",
       "      <td>NC_056147_NAD4_4</td>\n",
       "      <td>NC_056148_NAD4_2</td>\n",
       "      <td>NC_056195_NAD4_1</td>\n",
       "      <td>NC_061762_NAD4_1</td>\n",
       "      <td>NC_069201_NAD4_1</td>\n",
       "      <td>NC_070174_NAD4_1</td>\n",
       "      <td>NC_071210_NAD4_2</td>\n",
       "      <td>NC_082275_NAD4_1</td>\n",
       "      <td>NC_082276_NAD4_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>1.000</td>\n",
       "      <td>NC_015789_NAD4_2</td>\n",
       "      <td>NC_015991_NAD4_1</td>\n",
       "      <td>NC_023125_NAD4_2</td>\n",
       "      <td>NC_023126_NAD4_2</td>\n",
       "      <td>NC_023127_NAD4_2</td>\n",
       "      <td>NC_023128_NAD4_2</td>\n",
       "      <td>NC_025200_NAD4_2</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_NAD4_1</td>\n",
       "      <td>NC_056147_NAD4_1</td>\n",
       "      <td>NC_056148_NAD4_1</td>\n",
       "      <td>*</td>\n",
       "      <td>NC_061762_NAD4_2</td>\n",
       "      <td>NC_069201_NAD4_2</td>\n",
       "      <td>NC_070174_NAD4_2</td>\n",
       "      <td>NC_071210_NAD4_1</td>\n",
       "      <td>NC_082275_NAD4_2</td>\n",
       "      <td>NC_082276_NAD4_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>1.000</td>\n",
       "      <td>NC_015789_NAD5_1</td>\n",
       "      <td>*</td>\n",
       "      <td>NC_023125_NAD5_1</td>\n",
       "      <td>NC_023126_NAD5_1</td>\n",
       "      <td>NC_023127_NAD5_1</td>\n",
       "      <td>NC_023128_NAD5_1</td>\n",
       "      <td>NC_025200_NAD5_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_NAD5_1</td>\n",
       "      <td>NC_056147_NAD5_1</td>\n",
       "      <td>NC_056148_NAD5_1</td>\n",
       "      <td>NC_056195_NAD5_1</td>\n",
       "      <td>NC_061762_NAD5_1</td>\n",
       "      <td>NC_069201_NAD5_1</td>\n",
       "      <td>NC_070174_NAD5_1</td>\n",
       "      <td>NC_071210_NAD5_1</td>\n",
       "      <td>NC_082275_NAD5_1</td>\n",
       "      <td>NC_082276_NAD5_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>0.857</td>\n",
       "      <td>NC_015789_NAD3_1</td>\n",
       "      <td>NC_015991_NAD3_1</td>\n",
       "      <td>NC_023125_NAD3_1</td>\n",
       "      <td>NC_023126_NAD3_1</td>\n",
       "      <td>NC_023127_NAD3_1</td>\n",
       "      <td>NC_023128_NAD3_1</td>\n",
       "      <td>NC_025200_NAD3_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_NAD3_1</td>\n",
       "      <td>NC_056147_NAD3_1</td>\n",
       "      <td>NC_056148_NAD3_1</td>\n",
       "      <td>NC_056195_NAD3_1</td>\n",
       "      <td>NC_061762_NAD3_1</td>\n",
       "      <td>NC_069201_NAD3_1</td>\n",
       "      <td>NC_070174_NAD3_1</td>\n",
       "      <td>NC_071210_NAD3_1</td>\n",
       "      <td>NC_082275_NAD3_1</td>\n",
       "      <td>NC_082276_NAD3_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>0.498</td>\n",
       "      <td>NC_015789_COX1_1</td>\n",
       "      <td>*</td>\n",
       "      <td>NC_023125_COX1_1</td>\n",
       "      <td>NC_023126_COX1_1</td>\n",
       "      <td>NC_023127_COX1_1</td>\n",
       "      <td>NC_023128_COX1_1</td>\n",
       "      <td>NC_025200_COX1_1</td>\n",
       "      <td>...</td>\n",
       "      <td>NC_056146_COX1_1</td>\n",
       "      <td>NC_056147_COX1_1</td>\n",
       "      <td>NC_056148_COX1_1</td>\n",
       "      <td>NC_056195_COX1_1</td>\n",
       "      <td>NC_061762_COX1_1</td>\n",
       "      <td>NC_069201_COX1_1</td>\n",
       "      <td>NC_070174_COX1_1</td>\n",
       "      <td>NC_071210_COX1_1</td>\n",
       "      <td>NC_082275_COX1_1</td>\n",
       "      <td>NC_082276_COX1_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    # Species  Genes  Alg.-Conn. NC_015789.mt_prots.fa NC_015991.mt_prots.fa  \\\n",
       "0          22     22       1.000      NC_015789_ATP6_1      NC_015991_ATP6_1   \n",
       "1          22     22       1.000      NC_015789_COX2_1      NC_015991_COX2_1   \n",
       "2          22     22       1.000      NC_015789_COX3_1      NC_015991_COX3_1   \n",
       "3          22     22       1.000      NC_015789_CYTB_1      NC_015991_CYTB_1   \n",
       "4          22     22       0.909      NC_015789_NAD1_1      NC_015991_NAD1_1   \n",
       "5          22     22       0.909      NC_015789_NAD6_1      NC_015991_NAD6_1   \n",
       "6          21     21       1.000      NC_015789_ATP8_1                     *   \n",
       "7          21     21       1.000      NC_015789_NAD2_1                     *   \n",
       "8          21     21       1.000      NC_015789_NAD4_1                     *   \n",
       "9          21     21       1.000      NC_015789_NAD4_2      NC_015991_NAD4_1   \n",
       "10         21     21       1.000      NC_015789_NAD5_1                     *   \n",
       "11         21     21       0.857      NC_015789_NAD3_1      NC_015991_NAD3_1   \n",
       "12         20     20       0.498      NC_015789_COX1_1                     *   \n",
       "\n",
       "   NC_023125.mt_prots.fa NC_023126.mt_prots.fa NC_023127.mt_prots.fa  \\\n",
       "0       NC_023125_ATP6_1      NC_023126_ATP6_1      NC_023127_ATP6_1   \n",
       "1       NC_023125_COX2_1      NC_023126_COX2_1      NC_023127_COX2_1   \n",
       "2       NC_023125_COX3_1      NC_023126_COX3_1      NC_023127_COX3_1   \n",
       "3        NC_023125_COB_1       NC_023126_COB_1       NC_023127_COB_1   \n",
       "4       NC_023125_NAD1_1      NC_023126_NAD1_1      NC_023127_NAD1_1   \n",
       "5       NC_023125_NAD6_1      NC_023126_NAD6_1      NC_023127_NAD6_1   \n",
       "6       NC_023125_ATP8_1      NC_023126_ATP8_1      NC_023127_ATP8_1   \n",
       "7       NC_023125_NAD2_1      NC_023126_NAD2_1      NC_023127_NAD2_1   \n",
       "8       NC_023125_NAD4_1      NC_023126_NAD4_1      NC_023127_NAD4_1   \n",
       "9       NC_023125_NAD4_2      NC_023126_NAD4_2      NC_023127_NAD4_2   \n",
       "10      NC_023125_NAD5_1      NC_023126_NAD5_1      NC_023127_NAD5_1   \n",
       "11      NC_023125_NAD3_1      NC_023126_NAD3_1      NC_023127_NAD3_1   \n",
       "12      NC_023125_COX1_1      NC_023126_COX1_1      NC_023127_COX1_1   \n",
       "\n",
       "   NC_023128.mt_prots.fa NC_025200.mt_prots.fa  ... NC_056146.mt_prots.fa  \\\n",
       "0       NC_023128_ATP6_1      NC_025200_ATP6_1  ...      NC_056146_ATP6_1   \n",
       "1       NC_023128_COX2_1      NC_025200_COX2_1  ...      NC_056146_COX2_1   \n",
       "2       NC_023128_COX3_1      NC_025200_COX3_1  ...      NC_056146_COX3_1   \n",
       "3        NC_023128_COB_1       NC_025200_COB_1  ...       NC_056146_COB_1   \n",
       "4       NC_023128_NAD1_1      NC_025200_NAD1_1  ...      NC_056146_NAD1_1   \n",
       "5       NC_023128_NAD6_1      NC_025200_NAD6_1  ...      NC_056146_NAD6_1   \n",
       "6       NC_023128_ATP8_1      NC_025200_ATP8_1  ...      NC_056146_ATP8_1   \n",
       "7       NC_023128_NAD2_1      NC_025200_NAD2_1  ...      NC_056146_NAD2_1   \n",
       "8       NC_023128_NAD4_1      NC_025200_NAD4_1  ...      NC_056146_NAD4_4   \n",
       "9       NC_023128_NAD4_2      NC_025200_NAD4_2  ...      NC_056146_NAD4_1   \n",
       "10      NC_023128_NAD5_1      NC_025200_NAD5_1  ...      NC_056146_NAD5_1   \n",
       "11      NC_023128_NAD3_1      NC_025200_NAD3_1  ...      NC_056146_NAD3_1   \n",
       "12      NC_023128_COX1_1      NC_025200_COX1_1  ...      NC_056146_COX1_1   \n",
       "\n",
       "   NC_056147.mt_prots.fa NC_056148.mt_prots.fa NC_056195.mt_prots.fa  \\\n",
       "0       NC_056147_ATP6_1      NC_056148_ATP6_1      NC_056195_ATP6_1   \n",
       "1       NC_056147_COX2_1      NC_056148_COX2_1      NC_056195_COX2_1   \n",
       "2       NC_056147_COX3_1      NC_056148_COX3_1      NC_056195_COX3_1   \n",
       "3        NC_056147_COB_1       NC_056148_COB_1       NC_056195_COB_1   \n",
       "4       NC_056147_NAD1_1      NC_056148_NAD1_1      NC_056195_NAD1_1   \n",
       "5       NC_056147_NAD6_1      NC_056148_NAD6_1      NC_056195_NAD6_1   \n",
       "6       NC_056147_ATP8_1      NC_056148_ATP8_1      NC_056195_ATP8_1   \n",
       "7       NC_056147_NAD2_1      NC_056148_NAD2_1      NC_056195_NAD2_1   \n",
       "8       NC_056147_NAD4_4      NC_056148_NAD4_2      NC_056195_NAD4_1   \n",
       "9       NC_056147_NAD4_1      NC_056148_NAD4_1                     *   \n",
       "10      NC_056147_NAD5_1      NC_056148_NAD5_1      NC_056195_NAD5_1   \n",
       "11      NC_056147_NAD3_1      NC_056148_NAD3_1      NC_056195_NAD3_1   \n",
       "12      NC_056147_COX1_1      NC_056148_COX1_1      NC_056195_COX1_1   \n",
       "\n",
       "   NC_061762.mt_prots.fa NC_069201.mt_prots.fa NC_070174.mt_prots.fa  \\\n",
       "0       NC_061762_ATP6_1      NC_069201_ATP6_1      NC_070174_ATP6_1   \n",
       "1       NC_061762_COX2_1      NC_069201_COX2_1      NC_070174_COX2_1   \n",
       "2       NC_061762_COX3_1      NC_069201_COX3_1      NC_070174_COX3_1   \n",
       "3        NC_061762_COB_1       NC_069201_COB_1       NC_070174_COB_1   \n",
       "4       NC_061762_NAD1_1      NC_069201_NAD1_1      NC_070174_NAD1_1   \n",
       "5       NC_061762_NAD6_1      NC_069201_NAD6_1      NC_070174_NAD6_1   \n",
       "6       NC_061762_ATP8_1      NC_069201_ATP8_1      NC_070174_ATP8_1   \n",
       "7       NC_061762_NAD2_1      NC_069201_NAD2_1      NC_070174_NAD2_1   \n",
       "8       NC_061762_NAD4_1      NC_069201_NAD4_1      NC_070174_NAD4_1   \n",
       "9       NC_061762_NAD4_2      NC_069201_NAD4_2      NC_070174_NAD4_2   \n",
       "10      NC_061762_NAD5_1      NC_069201_NAD5_1      NC_070174_NAD5_1   \n",
       "11      NC_061762_NAD3_1      NC_069201_NAD3_1      NC_070174_NAD3_1   \n",
       "12      NC_061762_COX1_1      NC_069201_COX1_1      NC_070174_COX1_1   \n",
       "\n",
       "   NC_071210.mt_prots.fa NC_082275.mt_prots.fa NC_082276.mt_prots.fa  \n",
       "0       NC_071210_ATP6_1      NC_082275_ATP6_1      NC_082276_ATP6_1  \n",
       "1       NC_071210_COX2_1      NC_082275_COX2_1      NC_082276_COX2_1  \n",
       "2       NC_071210_COX3_1      NC_082275_COX3_1      NC_082276_COX3_1  \n",
       "3        NC_071210_COB_1       NC_082275_COB_1       NC_082276_COB_1  \n",
       "4       NC_071210_NAD1_1      NC_082275_NAD1_1      NC_082276_NAD1_1  \n",
       "5       NC_071210_NAD6_1      NC_082275_NAD6_1      NC_082276_NAD6_1  \n",
       "6       NC_071210_ATP8_1      NC_082275_ATP8_1      NC_082276_ATP8_1  \n",
       "7       NC_071210_NAD2_1      NC_082275_NAD2_1      NC_082276_NAD2_1  \n",
       "8       NC_071210_NAD4_2      NC_082275_NAD4_1      NC_082276_NAD4_1  \n",
       "9       NC_071210_NAD4_1      NC_082275_NAD4_2      NC_082276_NAD4_2  \n",
       "10      NC_071210_NAD5_1      NC_082275_NAD5_1      NC_082276_NAD5_1  \n",
       "11      NC_071210_NAD3_1      NC_082275_NAD3_1      NC_082276_NAD3_1  \n",
       "12      NC_071210_COX1_1      NC_082275_COX1_1      NC_082276_COX1_1  \n",
       "\n",
       "[13 rows x 25 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Good."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we shall drop every `*` from the dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all_cols_to_drop = df_all.columns[df_all.apply(lambda col: col.astype(str).str.contains(r'\\*', regex=True).any())]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If we drop `NC_015991.mt_prots.fa`, `NC_051483.mt_prots.fa`, `NC_056195.mt_prots.fa` columns there will be no `*` in our data! Meaning there will be a set of 13 genes with exactly the same set of taxa (PERFECT!!!)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['NC_015991.mt_prots.fa', 'NC_051483.mt_prots.fa',\n",
       "       'NC_056195.mt_prots.fa'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all_cols_to_drop"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Drop'em and save the df to `phylogenomics/protein_ortho_output/All.tsv` file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all = df_all.drop(columns=df_all_cols_to_drop)\n",
    "df_all.to_csv('phylogenomics/protein_ortho_output/All.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we shall start the process of extracting desired sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This script will write all the sequences families to be extracted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! Rscript scripts/write_family_names.R phylogenomics/protein_ortho_output/All.tsv phylogenomics/All/ phylogenomics/All_names/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "By this step we create a huge dataset of all the sequences we've been analysing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "! cat phylogenomics/Proteins_renamed_r3/*.fa > phylogenomics/all_pep.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a directory to store the desired sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "! mkdir phylogenomics/All_seqs/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, using `BBMap` once again we will extract desired sequences from `Proteinortho` output!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "for file in phylogenomics/All_names/family*.txt\n",
    "do \n",
    "  bbmap/filterbyname.sh in=phylogenomics/all_pep.fa out=phylogenomics/All_seqs/$(basename \"$file\" .names.txt).seq.fa include=t names=$file overwrite=true ignorejunk=true\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let us check the names of sequences in files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">NC_015789_ATP6_1\n"
     ]
    }
   ],
   "source": [
    "! head -1 phylogenomics/All_seqs/family1.seq.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Well, this is good. But for building the phylogenomic tree all the sequences in all the genes sets must be named the same! (e.g. not `>NC_015789_[GENE]`, but `>NC_015789`)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create the directory to store the renamed sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "! mkdir phylogenomics/All_seqs_renamed/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "for file in phylogenomics/All_seqs/*.fa\n",
    "do\n",
    "  # Process each file line by line\n",
    "  while read -r line; do\n",
    "    if [[ $line == \\>* ]]; then\n",
    "      # For all other headers, delete the second \"_\" and everything after it\n",
    "      echo \"$line\" | sed 's/\\(_[^_]*\\)\\(_.*\\)/\\1/' >> phylogenomics/All_seqs_renamed/$(basename \"$file\")\n",
    "    else\n",
    "      # For non-header lines (sequence data), copy them as-is\n",
    "      echo \"$line\" >> phylogenomics/All_seqs_renamed/$(basename \"$file\")\n",
    "    fi\n",
    "  done < \"$file\"\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">NC_015789\n"
     ]
    }
   ],
   "source": [
    "! head -1 phylogenomics/All_seqs_renamed/family1.seq.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Good."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we have everything ready for building phylogenomic tree!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create the directory to store MSAs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "mkdir phylogenomics/MSAs/\n",
    "mkdir phylogenomics/MSAs/All/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run `MAFFT`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "outputhat23=16\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8176 kb\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "tbfast-pair (aa) Version 7.525\n",
      "alg=L, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "outputhat23=16\n",
      "Loading 'hat3.seed' ... \n",
      "done.\n",
      "Writing hat3 for iterative refinement\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 1, compacttree = 0\n",
      "Constructing a UPGMA tree ... \n",
      "   10 / 19\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    18 /18 \n",
      "done.\n",
      "tbfast (aa) Version 7.525\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 16\n",
      "sueff_global = 0.100000\n",
      "nadd = 16\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   10 / 19\n",
      "Segment   1/  1    1- 274\n",
      "STEP 003-006-1  identical.   \n",
      "Oscillating.\n",
      "\n",
      "done\n",
      "dvtditr (aa) Version 7.525\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " L-INS-i (Probably most accurate, very slow)\n",
      " Iterative refinement method (<16) with LOCAL pairwise alignment information\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "outputhat23=16\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8176 kb\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "tbfast-pair (aa) Version 7.525\n",
      "alg=L, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "outputhat23=16\n",
      "Loading 'hat3.seed' ... \n",
      "done.\n",
      "Writing hat3 for iterative refinement\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 1, compacttree = 0\n",
      "Constructing a UPGMA tree ... \n",
      "   10 / 19\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    18 /18 \n",
      "done.\n",
      "tbfast (aa) Version 7.525\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 16\n",
      "sueff_global = 0.100000\n",
      "nadd = 16\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   10 / 19\n",
      "Segment   1/  1    1- 104\n",
      "STEP 002-017-1  identical.   \n",
      "Converged.\n",
      "\n",
      "done\n",
      "dvtditr (aa) Version 7.525\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " L-INS-i (Probably most accurate, very slow)\n",
      " Iterative refinement method (<16) with LOCAL pairwise alignment information\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "outputhat23=16\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8176 kb\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "tbfast-pair (aa) Version 7.525\n",
      "alg=L, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "outputhat23=16\n",
      "Loading 'hat3.seed' ... \n",
      "done.\n",
      "Writing hat3 for iterative refinement\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 1, compacttree = 0\n",
      "Constructing a UPGMA tree ... \n",
      "   10 / 19\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    18 /18 \n",
      "done.\n",
      "tbfast (aa) Version 7.525\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 16\n",
      "sueff_global = 0.100000\n",
      "nadd = 16\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   10 / 19\n",
      "Segment   1/  1    1-1083\n",
      "STEP 003-001-1  identical.    identical.    identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "done\n",
      "dvtditr (aa) Version 7.525\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " L-INS-i (Probably most accurate, very slow)\n",
      " Iterative refinement method (<16) with LOCAL pairwise alignment information\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "outputhat23=16\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8176 kb\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "tbfast-pair (aa) Version 7.525\n",
      "alg=L, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "outputhat23=16\n",
      "Loading 'hat3.seed' ... \n",
      "done.\n",
      "Writing hat3 for iterative refinement\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 1, compacttree = 0\n",
      "Constructing a UPGMA tree ... \n",
      "   10 / 19\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    18 /18 \n",
      "done.\n",
      "tbfast (aa) Version 7.525\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 16\n",
      "sueff_global = 0.100000\n",
      "nadd = 16\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   10 / 19\n",
      "Segment   1/  1    1- 178\n",
      "STEP 003-001-1  identical.   \n",
      "Oscillating.\n",
      "\n",
      "done\n",
      "dvtditr (aa) Version 7.525\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " L-INS-i (Probably most accurate, very slow)\n",
      " Iterative refinement method (<16) with LOCAL pairwise alignment information\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "outputhat23=16\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8176 kb\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "tbfast-pair (aa) Version 7.525\n",
      "alg=L, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "outputhat23=16\n",
      "Loading 'hat3.seed' ... \n",
      "done.\n",
      "Writing hat3 for iterative refinement\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 1, compacttree = 0\n",
      "Constructing a UPGMA tree ... \n",
      "   10 / 19\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    18 /18 \n",
      "done.\n",
      "tbfast (aa) Version 7.525\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 16\n",
      "sueff_global = 0.100000\n",
      "nadd = 16\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   10 / 19\n",
      "Segment   1/  1    1-1300\n",
      "STEP 003-003-0  identical.    rejected. rejected. identical.   \n",
      "Oscillating.\n",
      "\n",
      "done\n",
      "dvtditr (aa) Version 7.525\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " L-INS-i (Probably most accurate, very slow)\n",
      " Iterative refinement method (<16) with LOCAL pairwise alignment information\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "outputhat23=16\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8176 kb\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "tbfast-pair (aa) Version 7.525\n",
      "alg=L, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "outputhat23=16\n",
      "Loading 'hat3.seed' ... \n",
      "done.\n",
      "Writing hat3 for iterative refinement\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 1, compacttree = 0\n",
      "Constructing a UPGMA tree ... \n",
      "   10 / 19\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    18 /18 \n",
      "done.\n",
      "tbfast (aa) Version 7.525\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 16\n",
      "sueff_global = 0.100000\n",
      "nadd = 16\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   10 / 19\n",
      "Segment   1/  1    1- 253\n",
      "STEP 002-017-1  identical.   \n",
      "Converged.\n",
      "\n",
      "done\n",
      "dvtditr (aa) Version 7.525\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " L-INS-i (Probably most accurate, very slow)\n",
      " Iterative refinement method (<16) with LOCAL pairwise alignment information\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "outputhat23=16\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8176 kb\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "tbfast-pair (aa) Version 7.525\n",
      "alg=L, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "outputhat23=16\n",
      "Loading 'hat3.seed' ... \n",
      "done.\n",
      "Writing hat3 for iterative refinement\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 1, compacttree = 0\n",
      "Constructing a UPGMA tree ... \n",
      "   10 / 19\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    18 /18 \n",
      "done.\n",
      "tbfast (aa) Version 7.525\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 16\n",
      "sueff_global = 0.100000\n",
      "nadd = 16\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   10 / 19\n",
      "Segment   1/  1    1- 272\n",
      "STEP 003-001-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "done\n",
      "dvtditr (aa) Version 7.525\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " L-INS-i (Probably most accurate, very slow)\n",
      " Iterative refinement method (<16) with LOCAL pairwise alignment information\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "outputhat23=16\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8176 kb\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "tbfast-pair (aa) Version 7.525\n",
      "alg=L, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "outputhat23=16\n",
      "Loading 'hat3.seed' ... \n",
      "done.\n",
      "Writing hat3 for iterative refinement\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 1, compacttree = 0\n",
      "Constructing a UPGMA tree ... \n",
      "   10 / 19\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    18 /18 \n",
      "done.\n",
      "tbfast (aa) Version 7.525\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 16\n",
      "sueff_global = 0.100000\n",
      "nadd = 16\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   10 / 19\n",
      "Segment   1/  1    1- 401\n",
      "STEP 003-010-0  identical.    accepted.\n",
      "Converged.\n",
      "\n",
      "done\n",
      "dvtditr (aa) Version 7.525\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " L-INS-i (Probably most accurate, very slow)\n",
      " Iterative refinement method (<16) with LOCAL pairwise alignment information\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "outputhat23=16\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8176 kb\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "tbfast-pair (aa) Version 7.525\n",
      "alg=L, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "outputhat23=16\n",
      "Loading 'hat3.seed' ... \n",
      "done.\n",
      "Writing hat3 for iterative refinement\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 1, compacttree = 0\n",
      "Constructing a UPGMA tree ... \n",
      "   10 / 19\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    18 /18 \n",
      "done.\n",
      "tbfast (aa) Version 7.525\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 16\n",
      "sueff_global = 0.100000\n",
      "nadd = 16\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   10 / 19\n",
      "Segment   1/  1    1- 407\n",
      "STEP 003-006-0  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "done\n",
      "dvtditr (aa) Version 7.525\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " L-INS-i (Probably most accurate, very slow)\n",
      " Iterative refinement method (<16) with LOCAL pairwise alignment information\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "outputhat23=16\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8176 kb\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "tbfast-pair (aa) Version 7.525\n",
      "alg=L, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "outputhat23=16\n",
      "Loading 'hat3.seed' ... \n",
      "done.\n",
      "Writing hat3 for iterative refinement\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 1, compacttree = 0\n",
      "Constructing a UPGMA tree ... \n",
      "   10 / 19\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    18 /18 \n",
      "done.\n",
      "tbfast (aa) Version 7.525\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 16\n",
      "sueff_global = 0.100000\n",
      "nadd = 16\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   10 / 19\n",
      "Segment   1/  1    1- 305\n",
      "STEP 002-005-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "done\n",
      "dvtditr (aa) Version 7.525\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " L-INS-i (Probably most accurate, very slow)\n",
      " Iterative refinement method (<16) with LOCAL pairwise alignment information\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "outputhat23=16\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8176 kb\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "tbfast-pair (aa) Version 7.525\n",
      "alg=L, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "outputhat23=16\n",
      "Loading 'hat3.seed' ... \n",
      "done.\n",
      "Writing hat3 for iterative refinement\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 1, compacttree = 0\n",
      "Constructing a UPGMA tree ... \n",
      "   10 / 19\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    18 /18 \n",
      "done.\n",
      "tbfast (aa) Version 7.525\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 16\n",
      "sueff_global = 0.100000\n",
      "nadd = 16\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   10 / 19\n",
      "Segment   1/  1    1-  61\n",
      "STEP 002-017-1  identical.   \n",
      "Converged.\n",
      "\n",
      "done\n",
      "dvtditr (aa) Version 7.525\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " L-INS-i (Probably most accurate, very slow)\n",
      " Iterative refinement method (<16) with LOCAL pairwise alignment information\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "outputhat23=16\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8176 kb\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "tbfast-pair (aa) Version 7.525\n",
      "alg=L, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "outputhat23=16\n",
      "Loading 'hat3.seed' ... \n",
      "done.\n",
      "Writing hat3 for iterative refinement\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 1, compacttree = 0\n",
      "Constructing a UPGMA tree ... \n",
      "   10 / 19\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    18 /18 \n",
      "done.\n",
      "tbfast (aa) Version 7.525\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 16\n",
      "sueff_global = 0.100000\n",
      "nadd = 16\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   10 / 19\n",
      "Segment   1/  1    1- 633\n",
      "STEP 003-005-0  identical.    identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "done\n",
      "dvtditr (aa) Version 7.525\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " L-INS-i (Probably most accurate, very slow)\n",
      " Iterative refinement method (<16) with LOCAL pairwise alignment information\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "outputhat23=16\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8176 kb\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "tbfast-pair (aa) Version 7.525\n",
      "alg=L, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "outputhat23=16\n",
      "Loading 'hat3.seed' ... \n",
      "done.\n",
      "Writing hat3 for iterative refinement\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 1, compacttree = 0\n",
      "Constructing a UPGMA tree ... \n",
      "   10 / 19\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    18 /18 \n",
      "done.\n",
      "tbfast (aa) Version 7.525\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 16\n",
      "sueff_global = 0.100000\n",
      "nadd = 16\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   10 / 19\n",
      "Segment   1/  1    1- 500\n",
      "STEP 002-003-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "done\n",
      "dvtditr (aa) Version 7.525\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " L-INS-i (Probably most accurate, very slow)\n",
      " Iterative refinement method (<16) with LOCAL pairwise alignment information\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! for file in phylogenomics/All_seqs_renamed/*.fa; do mafft --auto \"$file\" > phylogenomics/MSAs/All/$(basename \"$file\" .fa).aln; done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a directory to store trimmed MSAs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "mkdir phylogenomics/trimmed_MSAs/\n",
    "mkdir phylogenomics/trimmed_MSAs/All/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run `trimAl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "! for file in phylogenomics/MSAs/All/*aln; do trimal -in $file -out phylogenomics/trimmed_MSAs/All/trimmed_$(basename \"$file\") -automated1; done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a directory to store `ModelFinder` data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "mkdir phylogenomics/model-finder/\n",
    "mkdir phylogenomics/model-finder/All/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run `ModelFinder`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IQ-TREE multicore version 2.4.0 for MacOS Intel 64-bit built Feb 12 2025\n",
      "Developed by Bui Quang Minh, Nguyen Lam Tung, Olga Chernomor, Heiko Schmidt,\n",
      "Dominik Schrempf, Michael Woodhams, Ly Trong Nhan, Thomas Wong\n",
      "\n",
      "Host:    Ilas-Mac-mini.local (SSE4.2, 16 GB RAM)\n",
      "Command: iqtree2 -m MFP -s phylogenomics/trimmed_MSAs/All/ --prefix phylogenomics/model-finder/All/All -T AUTO\n",
      "Seed:    594338 (Using SPRNG - Scalable Parallel Random Number Generator)\n",
      "Time:    Mon Mar 17 20:59:14 2025\n",
      "Kernel:  SSE2 - auto-detect threads (10 CPU cores detected)\n",
      "\n",
      "Reading 13 alignment files in directory phylogenomics/trimmed_MSAs/All/\n",
      "Reading alignment file phylogenomics/trimmed_MSAs/All/trimmed_family1.seq.aln ... Fasta format detected\n",
      "Reading fasta file: done in 0.000638962 secs using 62.13% CPU\n",
      "Alignment most likely contains protein sequences\n",
      "Alignment has 19 sequences with 258 columns, 131 distinct patterns\n",
      "73 parsimony-informative, 25 singleton sites, 160 constant sites\n",
      "           Gap/Ambiguity  Composition  p-value\n",
      "Analyzing sequences: done in 1.40667e-05 secs using 568.7% CPU\n",
      "   1  NC_015789    0.39%    passed    100.00%\n",
      "   2  NC_023125    0.00%    passed    100.00%\n",
      "   3  NC_023126    0.00%    passed    100.00%\n",
      "   4  NC_023127    0.00%    passed    100.00%\n",
      "   5  NC_023128    0.00%    passed    100.00%\n",
      "   6  NC_025200    1.16%    passed    100.00%\n",
      "   7  NC_027422    0.00%    passed    100.00%\n",
      "   8  NC_031375    0.39%    passed    100.00%\n",
      "   9  NC_033907   18.22%    passed     99.95%\n",
      "  10  NC_048507    0.00%    passed    100.00%\n",
      "  11  NC_056146    0.00%    passed    100.00%\n",
      "  12  NC_056147    0.00%    passed    100.00%\n",
      "  13  NC_056148    0.39%    passed    100.00%\n",
      "  14  NC_061762    0.00%    passed    100.00%\n",
      "  15  NC_069201    0.39%    passed    100.00%\n",
      "  16  NC_070174    0.00%    passed     99.95%\n",
      "  17  NC_071210    0.00%    passed    100.00%\n",
      "  18  NC_082275    0.00%    passed    100.00%\n",
      "  19  NC_082276    0.00%    passed    100.00%\n",
      "****  TOTAL        1.10%  0 sequences failed composition chi2 test (p-value<5%; df=19)\n",
      "Reading alignment file phylogenomics/trimmed_MSAs/All/trimmed_family10.seq.aln ... Fasta format detected\n",
      "Reading fasta file: done in 3.09944e-05 secs using 538.8% CPU\n",
      "Alignment most likely contains protein sequences\n",
      "Alignment has 19 sequences with 87 columns, 34 distinct patterns\n",
      "9 parsimony-informative, 9 singleton sites, 69 constant sites\n",
      "           Gap/Ambiguity  Composition  p-value\n",
      "Analyzing sequences: done in 5.00679e-06 secs using 359.5% CPU\n",
      "   1  NC_015789    0.00%    passed    100.00%\n",
      "   2  NC_023125    0.00%    passed    100.00%\n",
      "   3  NC_023126    0.00%    passed    100.00%\n",
      "   4  NC_023127    0.00%    passed    100.00%\n",
      "   5  NC_023128    0.00%    passed    100.00%\n",
      "   6  NC_025200    0.00%    passed    100.00%\n",
      "   7  NC_027422    0.00%    passed    100.00%\n",
      "   8  NC_031375    0.00%    passed    100.00%\n",
      "   9  NC_033907    0.00%    passed    100.00%\n",
      "  10  NC_048507    0.00%    passed    100.00%\n",
      "  11  NC_056146    0.00%    passed    100.00%\n",
      "  12  NC_056147    0.00%    passed    100.00%\n",
      "  13  NC_056148    0.00%    passed    100.00%\n",
      "  14  NC_061762    0.00%    passed    100.00%\n",
      "  15  NC_069201    0.00%    passed    100.00%\n",
      "  16  NC_070174    0.00%    passed     99.96%\n",
      "  17  NC_071210    0.00%    passed    100.00%\n",
      "  18  NC_082275    0.00%    passed    100.00%\n",
      "  19  NC_082276    0.00%    passed    100.00%\n",
      "****  TOTAL        0.00%  0 sequences failed composition chi2 test (p-value<5%; df=15)\n",
      "Reading alignment file phylogenomics/trimmed_MSAs/All/trimmed_family11.seq.aln ... Fasta format detected\n",
      "Reading fasta file: done in 0.000160217 secs using 904.4% CPU\n",
      "Alignment most likely contains protein sequences\n",
      "Alignment has 19 sequences with 653 columns, 233 distinct patterns\n",
      "150 parsimony-informative, 60 singleton sites, 442 constant sites\n",
      "           Gap/Ambiguity  Composition  p-value\n",
      "Analyzing sequences: done in 1.5974e-05 secs using 607.2% CPU\n",
      "   1  NC_015789    0.00%    passed    100.00%\n",
      "   2  NC_023125    0.00%    passed    100.00%\n",
      "   3  NC_023126    0.00%    passed    100.00%\n",
      "   4  NC_023127    0.00%    passed    100.00%\n",
      "   5  NC_023128    0.00%    passed    100.00%\n",
      "   6  NC_025200    0.00%    passed    100.00%\n",
      "   7  NC_027422    0.00%    passed    100.00%\n",
      "   8  NC_031375    0.00%    passed    100.00%\n",
      "   9  NC_033907    5.51%    passed    100.00%\n",
      "  10  NC_048507    0.00%    passed    100.00%\n",
      "  11  NC_056146    0.15%    passed    100.00%\n",
      "  12  NC_056147    0.00%    passed    100.00%\n",
      "  13  NC_056148    0.00%    passed    100.00%\n",
      "  14  NC_061762    0.00%    passed    100.00%\n",
      "  15  NC_069201    0.15%    passed     99.99%\n",
      "  16  NC_070174    0.00%    passed    100.00%\n",
      "  17  NC_071210    0.00%    passed    100.00%\n",
      "  18  NC_082275    0.00%    passed    100.00%\n",
      "  19  NC_082276    0.00%    passed    100.00%\n",
      "****  TOTAL        0.31%  0 sequences failed composition chi2 test (p-value<5%; df=19)\n",
      "Reading alignment file phylogenomics/trimmed_MSAs/All/trimmed_family12.seq.aln ... Fasta format detected\n",
      "Reading fasta file: done in 4.60148e-05 secs using 925.8% CPU\n",
      "Alignment most likely contains protein sequences\n",
      "Alignment has 19 sequences with 135 columns, 70 distinct patterns\n",
      "44 parsimony-informative, 3 singleton sites, 88 constant sites\n",
      "           Gap/Ambiguity  Composition  p-value\n",
      "Analyzing sequences: done in 9.05991e-06 secs using 883% CPU\n",
      "   1  NC_015789    0.00%    passed     99.96%\n",
      "   2  NC_023125    0.00%    passed     99.86%\n",
      "   3  NC_023126   12.59%    passed    100.00%\n",
      "   4  NC_023127   12.59%    passed    100.00%\n",
      "   5  NC_023128   12.59%    passed    100.00%\n",
      "   6  NC_025200    0.00%    passed    100.00%\n",
      "   7  NC_027422    0.00%    passed    100.00%\n",
      "   8  NC_031375    0.00%    passed    100.00%\n",
      "   9  NC_033907    7.41%    passed     99.97%\n",
      "  10  NC_048507    0.00%    passed    100.00%\n",
      "  11  NC_056146    0.00%    passed     99.97%\n",
      "  12  NC_056147    0.00%    passed     99.97%\n",
      "  13  NC_056148    0.00%    passed     99.97%\n",
      "  14  NC_061762    0.00%    passed    100.00%\n",
      "  15  NC_069201    0.00%    passed     99.62%\n",
      "  16  NC_070174    0.00%    passed     99.99%\n",
      "  17  NC_071210    0.00%    passed     99.97%\n",
      "  18  NC_082275    0.00%    passed    100.00%\n",
      "  19  NC_082276    0.00%    passed    100.00%\n",
      "****  TOTAL        2.38%  0 sequences failed composition chi2 test (p-value<5%; df=19)\n",
      "Reading alignment file phylogenomics/trimmed_MSAs/All/trimmed_family13.seq.aln ... Fasta format detected\n",
      "Reading fasta file: done in 0.000138044 secs using 909.9% CPU\n",
      "Alignment most likely contains protein sequences\n",
      "Alignment has 19 sequences with 526 columns, 138 distinct patterns\n",
      "64 parsimony-informative, 40 singleton sites, 422 constant sites\n",
      "           Gap/Ambiguity  Composition  p-value\n",
      "Analyzing sequences: done in 3.69549e-05 secs using 403.2% CPU\n",
      "   1  NC_015789    0.00%    passed    100.00%\n",
      "   2  NC_023125   11.03%    passed    100.00%\n",
      "   3  NC_023126   11.03%    passed    100.00%\n",
      "   4  NC_023127   11.03%    passed    100.00%\n",
      "   5  NC_023128   11.03%    passed    100.00%\n",
      "   6  NC_025200    0.00%    passed    100.00%\n",
      "   7  NC_027422    0.57%    passed    100.00%\n",
      "   8  NC_031375    0.00%    passed    100.00%\n",
      "   9  NC_033907    0.00%    passed    100.00%\n",
      "  10  NC_048507   11.03%    passed    100.00%\n",
      "  11  NC_056146    0.95%    passed    100.00%\n",
      "  12  NC_056147    0.95%    passed    100.00%\n",
      "  13  NC_056148    0.95%    passed    100.00%\n",
      "  14  NC_061762    0.19%    passed    100.00%\n",
      "  15  NC_069201    0.57%    passed    100.00%\n",
      "  16  NC_070174    0.00%    passed    100.00%\n",
      "  17  NC_071210    0.95%    passed    100.00%\n",
      "  18  NC_082275   11.03%    passed    100.00%\n",
      "  19  NC_082276   11.03%    passed    100.00%\n",
      "****  TOTAL        4.33%  0 sequences failed composition chi2 test (p-value<5%; df=19)\n",
      "Reading alignment file phylogenomics/trimmed_MSAs/All/trimmed_family2.seq.aln ... Fasta format detected\n",
      "Reading fasta file: done in 7.20024e-05 secs using 583.3% CPU\n",
      "Alignment most likely contains protein sequences\n",
      "Alignment has 19 sequences with 249 columns, 91 distinct patterns\n",
      "57 parsimony-informative, 11 singleton sites, 181 constant sites\n",
      "           Gap/Ambiguity  Composition  p-value\n",
      "Analyzing sequences: done in 7.15256e-06 secs using 461.4% CPU\n",
      "   1  NC_015789    0.00%    passed    100.00%\n",
      "   2  NC_023125    0.00%    passed    100.00%\n",
      "   3  NC_023126    0.00%    passed    100.00%\n",
      "   4  NC_023127    0.00%    passed    100.00%\n",
      "   5  NC_023128    0.00%    passed    100.00%\n",
      "   6  NC_025200    0.00%    passed    100.00%\n",
      "   7  NC_027422    0.00%    passed    100.00%\n",
      "   8  NC_031375    0.00%    passed    100.00%\n",
      "   9  NC_033907    7.23%    passed    100.00%\n",
      "  10  NC_048507    0.00%    passed    100.00%\n",
      "  11  NC_056146    0.00%    passed    100.00%\n",
      "  12  NC_056147    0.00%    passed    100.00%\n",
      "  13  NC_056148    0.00%    passed    100.00%\n",
      "  14  NC_061762    0.00%    passed    100.00%\n",
      "  15  NC_069201    0.00%    passed    100.00%\n",
      "  16  NC_070174    0.00%    passed    100.00%\n",
      "  17  NC_071210    0.00%    passed    100.00%\n",
      "  18  NC_082275    0.00%    passed    100.00%\n",
      "  19  NC_082276    0.00%    passed    100.00%\n",
      "****  TOTAL        0.38%  0 sequences failed composition chi2 test (p-value<5%; df=19)\n",
      "Reading alignment file phylogenomics/trimmed_MSAs/All/trimmed_family3.seq.aln ... Fasta format detected\n",
      "Reading fasta file: done in 0.00014019 secs using 857.4% CPU\n",
      "Alignment most likely contains protein sequences\n",
      "Alignment has 19 sequences with 254 columns, 116 distinct patterns\n",
      "68 parsimony-informative, 31 singleton sites, 155 constant sites\n",
      "           Gap/Ambiguity  Composition  p-value\n",
      "Analyzing sequences: done in 1.00136e-05 secs using 519.3% CPU\n",
      "   1  NC_015789    0.00%    passed    100.00%\n",
      "   2  NC_023125    0.00%    passed    100.00%\n",
      "   3  NC_023126    0.00%    passed    100.00%\n",
      "   4  NC_023127    0.00%    passed    100.00%\n",
      "   5  NC_023128    0.00%    passed    100.00%\n",
      "   6  NC_025200    0.00%    passed     99.85%\n",
      "   7  NC_027422    0.00%    passed    100.00%\n",
      "   8  NC_031375    0.00%    passed    100.00%\n",
      "   9  NC_033907    0.00%    passed    100.00%\n",
      "  10  NC_048507    0.00%    passed     99.35%\n",
      "  11  NC_056146    0.00%    passed    100.00%\n",
      "  12  NC_056147    0.00%    passed    100.00%\n",
      "  13  NC_056148    0.00%    passed    100.00%\n",
      "  14  NC_061762    0.00%    passed     99.97%\n",
      "  15  NC_069201    0.00%    passed    100.00%\n",
      "  16  NC_070174    0.00%    passed    100.00%\n",
      "  17  NC_071210    0.00%    passed    100.00%\n",
      "  18  NC_082275    0.00%    passed     98.76%\n",
      "  19  NC_082276    0.00%    passed     99.35%\n",
      "****  TOTAL        0.00%  0 sequences failed composition chi2 test (p-value<5%; df=19)\n",
      "Reading alignment file phylogenomics/trimmed_MSAs/All/trimmed_family4.seq.aln ... Fasta format detected\n",
      "Reading fasta file: done in 0.000103951 secs using 633% CPU\n",
      "Alignment most likely contains protein sequences\n",
      "Alignment has 19 sequences with 388 columns, 134 distinct patterns\n",
      "81 parsimony-informative, 36 singleton sites, 271 constant sites\n",
      "           Gap/Ambiguity  Composition  p-value\n",
      "Analyzing sequences: done in 1.3113e-05 secs using 533.8% CPU\n",
      "   1  NC_015789    0.00%    passed    100.00%\n",
      "   2  NC_023125    0.00%    passed    100.00%\n",
      "   3  NC_023126    0.00%    passed    100.00%\n",
      "   4  NC_023127    0.52%    passed    100.00%\n",
      "   5  NC_023128    0.00%    passed    100.00%\n",
      "   6  NC_025200    0.00%    passed    100.00%\n",
      "   7  NC_027422    0.00%    passed    100.00%\n",
      "   8  NC_031375    0.00%    passed    100.00%\n",
      "   9  NC_033907    0.52%    passed    100.00%\n",
      "  10  NC_048507    0.00%    passed    100.00%\n",
      "  11  NC_056146    0.00%    passed    100.00%\n",
      "  12  NC_056147    0.00%    passed    100.00%\n",
      "  13  NC_056148    0.00%    passed    100.00%\n",
      "  14  NC_061762    0.00%    passed    100.00%\n",
      "  15  NC_069201    0.00%    passed    100.00%\n",
      "  16  NC_070174    0.00%    passed    100.00%\n",
      "  17  NC_071210    0.00%    passed    100.00%\n",
      "  18  NC_082275    0.00%    passed    100.00%\n",
      "  19  NC_082276    0.00%    passed    100.00%\n",
      "****  TOTAL        0.05%  0 sequences failed composition chi2 test (p-value<5%; df=19)\n",
      "Reading alignment file phylogenomics/trimmed_MSAs/All/trimmed_family5.seq.aln ... Fasta format detected\n",
      "Reading fasta file: done in 8.98838e-05 secs using 672% CPU\n",
      "Alignment most likely contains protein sequences\n",
      "Alignment has 19 sequences with 360 columns, 162 distinct patterns\n",
      "92 parsimony-informative, 46 singleton sites, 222 constant sites\n",
      "           Gap/Ambiguity  Composition  p-value\n",
      "Analyzing sequences: done in 2.5034e-05 secs using 327.6% CPU\n",
      "   1  NC_015789    0.00%    passed    100.00%\n",
      "   2  NC_023125    0.00%    passed    100.00%\n",
      "   3  NC_023126    0.00%    passed    100.00%\n",
      "   4  NC_023127    0.00%    passed    100.00%\n",
      "   5  NC_023128    0.00%    passed    100.00%\n",
      "   6  NC_025200    0.00%    passed     99.95%\n",
      "   7  NC_027422    0.00%    passed    100.00%\n",
      "   8  NC_031375    0.00%    passed    100.00%\n",
      "   9  NC_033907    1.67%    passed    100.00%\n",
      "  10  NC_048507    0.00%    passed    100.00%\n",
      "  11  NC_056146    0.00%    passed    100.00%\n",
      "  12  NC_056147    0.00%    passed    100.00%\n",
      "  13  NC_056148    0.28%    passed     99.98%\n",
      "  14  NC_061762    0.00%    passed     99.97%\n",
      "  15  NC_069201    0.00%    passed    100.00%\n",
      "  16  NC_070174    0.00%    passed    100.00%\n",
      "  17  NC_071210    0.00%    passed    100.00%\n",
      "  18  NC_082275    0.00%    passed    100.00%\n",
      "  19  NC_082276    0.00%    passed    100.00%\n",
      "****  TOTAL        0.10%  0 sequences failed composition chi2 test (p-value<5%; df=19)\n",
      "Reading alignment file phylogenomics/trimmed_MSAs/All/trimmed_family6.seq.aln ... Fasta format detected\n",
      "Reading fasta file: done in 0.000145912 secs using 442.7% CPU\n",
      "Alignment most likely contains protein sequences\n",
      "Alignment has 19 sequences with 223 columns, 165 distinct patterns\n",
      "128 parsimony-informative, 16 singleton sites, 79 constant sites\n",
      "           Gap/Ambiguity  Composition  p-value\n",
      "Analyzing sequences: done in 3.60012e-05 secs using 430.5% CPU\n",
      "   1  NC_015789    0.00%    passed     99.44%\n",
      "   2  NC_023125    0.00%    passed    100.00%\n",
      "   3  NC_023126    0.00%    passed    100.00%\n",
      "   4  NC_023127    0.00%    passed    100.00%\n",
      "   5  NC_023128    0.00%    passed    100.00%\n",
      "   6  NC_025200    1.79%    passed     98.88%\n",
      "   7  NC_027422    5.83%    passed     89.07%\n",
      "   8  NC_031375    2.24%    passed     99.89%\n",
      "   9  NC_033907   15.25%    passed     90.60%\n",
      "  10  NC_048507    2.24%    passed     99.66%\n",
      "  11  NC_056146    0.45%    passed     99.80%\n",
      "  12  NC_056147    0.45%    passed     99.80%\n",
      "  13  NC_056148    0.45%    passed     99.98%\n",
      "  14  NC_061762    1.35%    passed     94.09%\n",
      "  15  NC_069201   16.14%    passed     72.40%\n",
      "  16  NC_070174    2.69%    passed     97.40%\n",
      "  17  NC_071210    0.45%    passed     99.75%\n",
      "  18  NC_082275    2.24%    passed     99.05%\n",
      "  19  NC_082276    2.24%    passed     99.66%\n",
      "****  TOTAL        2.83%  0 sequences failed composition chi2 test (p-value<5%; df=19)\n",
      "Reading alignment file phylogenomics/trimmed_MSAs/All/trimmed_family7.seq.aln ... Fasta format detected\n",
      "Reading fasta file: done in 5.6982e-05 secs using 251% CPU\n",
      "Alignment most likely contains protein sequences\n",
      "Alignment has 19 sequences with 48 columns, 37 distinct patterns\n",
      "15 parsimony-informative, 11 singleton sites, 22 constant sites\n",
      "           Gap/Ambiguity  Composition  p-value\n",
      "Analyzing sequences: done in 1.00136e-05 secs using 639.1% CPU\n",
      "   1  NC_015789    0.00%    passed     99.99%\n",
      "   2  NC_023125    0.00%    passed     99.98%\n",
      "   3  NC_023126    0.00%    passed     99.98%\n",
      "   4  NC_023127    0.00%    passed     99.97%\n",
      "   5  NC_023128    0.00%    passed     99.98%\n",
      "   6  NC_025200    0.00%    passed    100.00%\n",
      "   7  NC_027422    0.00%    passed    100.00%\n",
      "   8  NC_031375    0.00%    passed     99.97%\n",
      "   9  NC_033907    0.00%    passed    100.00%\n",
      "  10  NC_048507    0.00%    passed    100.00%\n",
      "  11  NC_056146    0.00%    passed    100.00%\n",
      "  12  NC_056147    0.00%    passed    100.00%\n",
      "  13  NC_056148    0.00%    passed     99.95%\n",
      "  14  NC_061762    0.00%    passed    100.00%\n",
      "  15  NC_069201    0.00%    passed     13.95%\n",
      "  16  NC_070174    0.00%    passed     99.97%\n",
      "  17  NC_071210    0.00%    passed    100.00%\n",
      "  18  NC_082275    0.00%    passed     99.97%\n",
      "  19  NC_082276    0.00%    passed    100.00%\n",
      "****  TOTAL        0.00%  0 sequences failed composition chi2 test (p-value<5%; df=14)\n",
      "Reading alignment file phylogenomics/trimmed_MSAs/All/trimmed_family8.seq.aln ... Fasta format detected\n",
      "Reading fasta file: done in 0.000272989 secs using 531.2% CPU\n",
      "Alignment most likely contains protein sequences\n",
      "Alignment has 19 sequences with 552 columns, 326 distinct patterns\n",
      "231 parsimony-informative, 68 singleton sites, 253 constant sites\n",
      "           Gap/Ambiguity  Composition  p-value\n",
      "Analyzing sequences: done in 2.59876e-05 secs using 423.3% CPU\n",
      "   1  NC_015789    0.00%    passed     99.75%\n",
      "   2  NC_023125    0.00%    passed     99.97%\n",
      "   3  NC_023126    0.00%    passed     99.97%\n",
      "   4  NC_023127    0.00%    passed     99.95%\n",
      "   5  NC_023128    0.00%    passed     99.97%\n",
      "   6  NC_025200    0.00%    passed     99.97%\n",
      "   7  NC_027422    0.00%    passed     99.53%\n",
      "   8  NC_031375    0.00%    passed     99.93%\n",
      "   9  NC_033907   22.46%    passed     94.84%\n",
      "  10  NC_048507    0.00%    passed     99.98%\n",
      "  11  NC_056146    0.00%    passed     99.47%\n",
      "  12  NC_056147    0.36%    passed     98.32%\n",
      "  13  NC_056148    0.18%    passed     98.04%\n",
      "  14  NC_061762    0.00%    passed     99.94%\n",
      "  15  NC_069201    0.36%    passed     85.13%\n",
      "  16  NC_070174   14.31%    passed     96.04%\n",
      "  17  NC_071210    0.36%    passed     98.54%\n",
      "  18  NC_082275    0.00%    passed     99.99%\n",
      "  19  NC_082276    0.00%    passed     99.98%\n",
      "****  TOTAL        2.00%  0 sequences failed composition chi2 test (p-value<5%; df=19)\n",
      "Reading alignment file phylogenomics/trimmed_MSAs/All/trimmed_family9.seq.aln ... Fasta format detected\n",
      "Reading fasta file: done in 0.000288963 secs using 366.5% CPU\n",
      "Alignment most likely contains protein sequences\n",
      "Alignment has 19 sequences with 489 columns, 214 distinct patterns\n",
      "135 parsimony-informative, 58 singleton sites, 296 constant sites\n",
      "           Gap/Ambiguity  Composition  p-value\n",
      "Analyzing sequences: done in 5.00679e-06 secs using 739% CPU\n",
      "   1  NC_015789    0.41%    passed    100.00%\n",
      "   2  NC_023125    0.41%    passed    100.00%\n",
      "   3  NC_023126    0.41%    passed    100.00%\n",
      "   4  NC_023127    0.41%    passed    100.00%\n",
      "   5  NC_023128    0.41%    passed    100.00%\n",
      "   6  NC_025200    0.00%    passed     99.65%\n",
      "   7  NC_027422    0.00%    passed     99.92%\n",
      "   8  NC_031375    0.41%    passed     99.98%\n",
      "   9  NC_033907   15.75%    passed     99.97%\n",
      "  10  NC_048507    0.00%    passed     99.99%\n",
      "  11  NC_056146    1.84%    passed     99.97%\n",
      "  12  NC_056147    1.84%    passed     99.97%\n",
      "  13  NC_056148    1.84%    passed     99.99%\n",
      "  14  NC_061762    0.00%    passed    100.00%\n",
      "  15  NC_069201    0.20%    passed     99.74%\n",
      "  16  NC_070174    0.00%    passed     99.99%\n",
      "  17  NC_071210    1.02%    passed     99.97%\n",
      "  18  NC_082275    0.00%    passed     99.99%\n",
      "  19  NC_082276    0.00%    passed     99.99%\n",
      "****  TOTAL        1.31%  0 sequences failed composition chi2 test (p-value<5%; df=19)\n",
      "NOTE: NC_082276 is identical to NC_048507 but kept for subsequent analysis\n",
      "\n",
      "\n",
      "Create initial parsimony tree by phylogenetic likelihood library (PLL)... 0.003 seconds\n",
      "Measuring multi-threading efficiency up to 10 CPU cores\n",
      "Increase to 10 rounds for branch lengths\n",
      "33 trees examined\n",
      "Threads: 1 / Time: 10.101 sec / Speedup: 1.000 / Efficiency: 100% / LogL: -42480\n",
      "Threads: 2 / Time: 6.644 sec / Speedup: 1.520 / Efficiency: 76% / LogL: -42480\n",
      "Threads: 3 / Time: 4.671 sec / Speedup: 2.162 / Efficiency: 72% / LogL: -42480\n",
      "Threads: 4 / Time: 3.948 sec / Speedup: 2.558 / Efficiency: 64% / LogL: -42480\n",
      "Threads: 5 / Time: 3.971 sec / Speedup: 2.544 / Efficiency: 51% / LogL: -42480\n",
      "Threads: 6 / Time: 3.651 sec / Speedup: 2.767 / Efficiency: 46% / LogL: -42480\n",
      "BEST NUMBER OF THREADS: 4\n",
      "\n",
      "Perform fast likelihood tree search using LG+I+G model...\n",
      "Estimate model parameters (epsilon = 5.000)\n",
      "Perform nearest neighbor interchange...\n",
      "Estimate model parameters (epsilon = 1.000)\n",
      "1. Initial log-likelihood: -33287.259\n",
      "Optimal log-likelihood: -33287.088\n",
      "Proportion of invariable sites: 0.323\n",
      "Gamma shape alpha: 0.527\n",
      "Parameters optimization took 1 rounds (0.047 sec)\n",
      "Time for fast ML tree search: 0.294 seconds\n",
      "\n",
      "NOTE: ModelFinder requires 42 MB RAM!\n",
      "ModelFinder will test up to 1232 protein models (sample size: 4222 epsilon: 0.100) ...\n",
      " No. Model         -LnL         df  AIC          AICc         BIC\n",
      "  1  LG            35050.530    35  70171.060    70171.662    70393.242\n",
      "  2  LG+I          33684.321    36  67440.642    67441.278    67669.172\n",
      "  3  LG+G4         33340.418    36  66752.836    66753.472    66981.366\n",
      "  4  LG+I+G4       33286.443    37  66646.886    66647.558    66881.764\n",
      "  5  LG+R2         33340.621    37  66755.243    66755.915    66990.121\n",
      "  6  LG+R3         33223.817    39  66525.633    66526.379    66773.208\n",
      "  7  LG+R4         33222.255    41  66526.511    66527.334    66786.781\n",
      " 14  LG+I+R2       33239.608    38  66555.216    66555.925    66796.443\n",
      " 15  LG+I+R3       33220.753    40  66521.506    66522.291    66775.429\n",
      " 16  LG+I+R4       33220.537    42  66525.074    66525.939    66791.693\n",
      " 28  LG+F+R3       32520.964    58  65157.928    65159.572    65526.116\n",
      " 37  LG+F+I+R3     32515.300    59  65148.600    65150.301    65523.135\n",
      " 50  WAG+R3        33422.534    39  66923.068    66923.814    67170.642\n",
      " 59  WAG+I+R3      33421.512    40  66923.024    66923.809    67176.947\n",
      " 72  WAG+F+R3      32522.561    58  65161.122    65162.766    65529.310\n",
      " 81  WAG+F+I+R3    32519.501    59  65157.003    65158.704    65531.539\n",
      " 94  JTT+R3        33346.248    39  66770.497    66771.243    67018.071\n",
      "103  JTT+I+R3      33344.076    40  66768.153    66768.937    67022.075\n",
      "116  JTT+F+R3      32520.074    58  65156.147    65157.791    65524.335\n",
      "125  JTT+F+I+R3    32516.722    59  65151.444    65153.145    65525.979\n",
      "WARNING: Normalizing state frequencies so that sum of them equals to 1\n",
      "138  Q.pfam+R3     33234.735    39  66547.469    66548.215    66795.044\n",
      "WARNING: Normalizing state frequencies so that sum of them equals to 1\n",
      "147  Q.pfam+I+R3   33231.608    40  66543.217    66544.001    66797.139\n",
      "WARNING: Normalizing state frequencies so that sum of them equals to 1\n",
      "160  Q.pfam+F+R3   32538.559    58  65193.119    65194.763    65561.306\n",
      "WARNING: Normalizing state frequencies so that sum of them equals to 1\n",
      "169  Q.pfam+F+I+R3 32532.154    59  65182.307    65184.008    65556.843\n",
      "WARNING: Normalizing state frequencies so that sum of them equals to 1\n",
      "182  Q.bird+R3     33694.848    39  67467.696    67468.442    67715.270\n",
      "WARNING: Normalizing state frequencies so that sum of them equals to 1\n",
      "191  Q.bird+I+R3   33688.149    40  67456.297    67457.082    67710.220\n",
      "WARNING: Normalizing state frequencies so that sum of them equals to 1\n",
      "204  Q.bird+F+R3   32770.497    58  65656.993    65658.637    66025.181\n",
      "WARNING: Normalizing state frequencies so that sum of them equals to 1\n",
      "213  Q.bird+F+I+R3 32760.901    59  65639.802    65641.503    66014.338\n",
      "WARNING: Normalizing state frequencies so that sum of them equals to 1\n",
      "226  Q.mammal+R3   33646.939    39  67371.878    67372.624    67619.452\n",
      "WARNING: Normalizing state frequencies so that sum of them equals to 1\n",
      "235  Q.mammal+I+R3 33640.678    40  67361.355    67362.140    67615.278\n",
      "WARNING: Normalizing state frequencies so that sum of them equals to 1\n",
      "248  Q.mammal+F+R3 32689.696    58  65495.393    65497.037    65863.581\n",
      "WARNING: Normalizing state frequencies so that sum of them equals to 1\n",
      "257  Q.mammal+F+I+R3 32680.989    59  65479.979    65481.680    65854.514\n",
      "270  Q.insect+R3   33141.610    39  66361.220    66361.966    66608.795\n",
      "279  Q.insect+I+R3 33137.566    40  66355.131    66355.916    66609.054\n",
      "292  Q.insect+F+R3 32543.004    58  65202.009    65203.653    65570.196\n",
      "301  Q.insect+F+I+R3 32537.307    59  65192.614    65194.315    65567.150\n",
      "WARNING: Normalizing state frequencies so that sum of them equals to 1\n",
      "314  Q.plant+R3    33070.973    39  66219.946    66220.692    66467.520\n",
      "WARNING: Normalizing state frequencies so that sum of them equals to 1\n",
      "323  Q.plant+I+R3  33064.893    40  66209.786    66210.570    66463.708\n",
      "WARNING: Normalizing state frequencies so that sum of them equals to 1\n",
      "336  Q.plant+F+R3  32481.255    58  65078.511    65080.155    65446.698\n",
      "WARNING: Normalizing state frequencies so that sum of them equals to 1\n",
      "345  Q.plant+F+I+R3 32474.121    59  65066.241    65067.942    65440.777\n",
      "WARNING: Normalizing state frequencies so that sum of them equals to 1\n",
      "358  Q.yeast+R3    33052.145    39  66182.291    66183.037    66429.865\n",
      "WARNING: Normalizing state frequencies so that sum of them equals to 1\n",
      "367  Q.yeast+I+R3  33048.636    40  66177.273    66178.057    66431.195\n",
      "WARNING: Normalizing state frequencies so that sum of them equals to 1\n",
      "380  Q.yeast+F+R3  32518.820    58  65153.640    65155.284    65521.828\n",
      "WARNING: Normalizing state frequencies so that sum of them equals to 1\n",
      "389  Q.yeast+F+I+R3 32514.002    59  65146.003    65147.704    65520.539\n",
      "402  JTTDCMut+R3   33355.382    39  66788.765    66789.511    67036.339\n",
      "411  JTTDCMut+I+R3 33353.415    40  66786.830    66787.614    67040.752\n",
      "424  JTTDCMut+F+R3 32517.437    58  65150.874    65152.518    65519.062\n",
      "433  JTTDCMut+F+I+R3 32514.393    59  65146.786    65148.487    65521.321\n",
      "446  DCMut+R3      33848.826    39  67775.652    67776.398    68023.227\n",
      "455  DCMut+I+R3    33845.993    40  67771.987    67772.771    68025.909\n",
      "468  DCMut+F+R3    32651.306    58  65418.613    65420.257    65786.800\n",
      "477  DCMut+F+I+R3  32647.654    59  65413.307    65415.008    65787.843\n",
      "490  VT+R3         33355.948    39  66789.896    66790.642    67037.471\n",
      "499  VT+I+R3       33355.825    40  66791.650    66792.435    67045.573\n",
      "512  VT+F+R3       32615.382    58  65346.764    65348.408    65714.952\n",
      "521  VT+F+I+R3     32613.435    59  65344.871    65346.572    65719.407\n",
      "534  PMB+R3        33441.910    39  66961.819    66962.565    67209.394\n",
      "543  PMB+I+R3      33441.221    40  66962.441    66963.226    67216.364\n",
      "556  PMB+F+R3      32797.605    58  65711.209    65712.853    66079.397\n",
      "565  PMB+F+I+R3    32794.764    59  65707.528    65709.229    66082.064\n",
      "578  Blosum62+R3   33468.647    39  67015.295    67016.041    67262.869\n",
      "587  Blosum62+I+R3 33468.589    40  67017.177    67017.962    67271.100\n",
      "600  Blosum62+F+R3 32805.466    58  65726.933    65728.577    66095.121\n",
      "609  Blosum62+F+I+R3 32803.072    59  65724.145    65725.846    66098.681\n",
      "622  Dayhoff+R3    33847.178    39  67772.355    67773.101    68019.930\n",
      "631  Dayhoff+I+R3  33844.114    40  67768.227    67769.012    68022.150\n",
      "644  Dayhoff+F+R3  32649.281    58  65414.563    65416.207    65782.751\n",
      "653  Dayhoff+F+I+R3 32645.514    59  65409.027    65410.728    65783.563\n",
      "666  mtREV+R3      33150.438    39  66378.876    66379.622    66626.451\n",
      "675  mtREV+I+R3    33143.283    40  66366.567    66367.351    66620.489\n",
      "688  mtREV+F+R3    32577.114    58  65270.228    65271.872    65638.416\n",
      "697  mtREV+F+I+R3  32568.916    59  65255.831    65257.532    65630.367\n",
      "710  mtART+R3      32936.684    39  65951.367    65952.113    66198.942\n",
      "719  mtART+I+R3    32926.311    40  65932.623    65933.407    66186.545\n",
      "732  mtART+F+R3    32663.344    58  65442.688    65444.332    65810.875\n",
      "741  mtART+F+I+R3  32650.742    59  65419.484    65421.185    65794.020\n",
      "754  mtZOA+R3      32731.851    39  65541.701    65542.447    65789.276\n",
      "763  mtZOA+I+R3    32725.298    40  65530.597    65531.381    65784.519\n",
      "776  mtZOA+F+R3    32481.691    58  65079.381    65081.025    65447.569\n",
      "785  mtZOA+F+I+R3  32473.224    59  65064.448    65066.149    65438.983\n",
      "798  mtMet+R3      32864.669    39  65807.337    65808.083    66054.912\n",
      "807  mtMet+I+R3    32858.712    40  65797.423    65798.208    66051.346\n",
      "820  mtMet+F+R3    32520.178    58  65156.356    65158.000    65524.544\n",
      "829  mtMet+F+I+R3  32513.657    59  65145.313    65147.014    65519.849\n",
      "842  mtVer+R3      33346.702    39  66771.405    66772.151    67018.979\n",
      "851  mtVer+I+R3    33339.690    40  66759.381    66760.165    67013.303\n",
      "864  mtVer+F+R3    32763.087    58  65642.175    65643.819    66010.362\n",
      "873  mtVer+F+I+R3  32752.777    59  65623.554    65625.255    65998.090\n",
      "886  mtInv+R3      32947.396    39  65972.791    65973.538    66220.366\n",
      "895  mtInv+I+R3    32942.177    40  65964.355    65965.139    66218.277\n",
      "908  mtInv+F+R3    32427.382    58  64970.764    64972.408    65338.952\n",
      "917  mtInv+F+I+R3  32423.137    59  64964.274    64965.975    65338.810\n",
      "930  mtMAM+R3      33459.310    39  66996.620    66997.366    67244.194\n",
      "939  mtMAM+I+R3    33446.230    40  66972.461    66973.245    67226.383\n",
      "952  mtMAM+F+R3    32974.369    58  66064.738    66066.382    66432.926\n",
      "961  mtMAM+F+I+R3  32954.134    59  66026.269    66027.970    66400.804\n",
      "WARNING: Normalizing state frequencies so that sum of them equals to 1\n",
      "974  FLAVI+R3      33946.119    39  67970.237    67970.983    68217.812\n",
      "WARNING: Normalizing state frequencies so that sum of them equals to 1\n",
      "983  FLAVI+I+R3    33935.788    40  67951.576    67952.361    68205.499\n",
      "WARNING: Normalizing state frequencies so that sum of them equals to 1\n",
      "996  FLAVI+F+R3    33195.958    58  66507.915    66509.559    66876.103\n",
      "WARNING: Normalizing state frequencies so that sum of them equals to 1\n",
      "1005  FLAVI+F+I+R3  33184.349    59  66486.699    66488.400    66861.235\n",
      "1018  HIVb+R3       33764.738    39  67607.476    67608.222    67855.050\n",
      "1027  HIVb+I+R3     33761.291    40  67602.582    67603.367    67856.505\n",
      "1040  HIVb+F+R3     32836.620    58  65789.239    65790.883    66157.427\n",
      "1049  HIVb+F+I+R3   32825.418    59  65768.837    65770.538    66143.372\n",
      "1062  HIVw+R3       34561.034    39  69200.068    69200.814    69447.643\n",
      "1071  HIVw+I+R3     34558.497    40  69196.994    69197.779    69450.917\n",
      "1084  HIVw+F+R3     33357.531    58  66831.063    66832.707    67199.250\n",
      "1093  HIVw+F+I+R3   33352.641    59  66823.283    66824.984    67197.819\n",
      "1106  FLU+R3        33578.114    39  67234.228    67234.974    67481.802\n",
      "1115  FLU+I+R3      33574.848    40  67229.695    67230.480    67483.618\n",
      "1128  FLU+F+R3      32789.802    58  65695.603    65697.247    66063.791\n",
      "1137  FLU+F+I+R3    32780.606    59  65679.212    65680.913    66053.747\n",
      "1150  rtREV+R3      33754.582    39  67587.163    67587.909    67834.738\n",
      "1159  rtREV+I+R3    33753.540    40  67587.080    67587.865    67841.003\n",
      "1172  rtREV+F+R3    32736.382    58  65588.763    65590.407    65956.951\n",
      "1181  rtREV+F+I+R3  32731.534    59  65581.069    65582.770    65955.605\n",
      "1194  cpREV+R3      32952.198    39  65982.396    65983.142    66229.971\n",
      "1203  cpREV+I+R3    32952.461    40  65984.923    65985.707    66238.845\n",
      "1216  cpREV+F+R3    32411.668    58  64939.336    64940.980    65307.524\n",
      "1225  cpREV+F+I+R3  32409.982    59  64937.964    64939.665    65312.500\n",
      "Akaike Information Criterion:           cpREV+F+I+R3\n",
      "Corrected Akaike Information Criterion: cpREV+F+I+R3\n",
      "Bayesian Information Criterion:         cpREV+F+R3\n",
      "Best-fit model: cpREV+F+R3 chosen according to BIC\n",
      "\n",
      "All model information printed to phylogenomics/model-finder/All/All.model.gz\n",
      "CPU time for ModelFinder: 222.113 seconds (0h:3m:42s)\n",
      "Wall-clock time for ModelFinder: 66.491 seconds (0h:1m:6s)\n",
      "\n",
      "NOTE: 12 MB RAM (0 GB) is required!\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "1. Initial log-likelihood: -32411.674\n",
      "Optimal log-likelihood: -32411.659\n",
      "Site proportion and rates:  (0.674,0.123) (0.254,1.676) (0.071,6.875)\n",
      "Parameters optimization took 1 rounds (0.071 sec)\n",
      "Wrote distance file to... \n",
      "Computing ML distances based on estimated model parameters...\n",
      "Calculating distance matrix: done in 0.00537395 secs using 326.7% CPU\n",
      "Computing ML distances took 0.005479 sec (of wall-clock time) 0.017895 sec (of CPU time)\n",
      "Setting up auxiliary I and S matrices: done in 4.00543e-05 secs using 369.5% CPU\n",
      "Computing RapidNJ tree took 0.000415 sec (of wall-clock time) 0.001594 sec (of CPU time)\n",
      "Log-likelihood of RapidNJ tree: -32442.932\n",
      "--------------------------------------------------------------------\n",
      "|             INITIALIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Generating 98 parsimony trees... 0.233 second\n",
      "Computing log-likelihood of 98 initial trees ... 1.013 seconds\n",
      "Current best score: -32411.659\n",
      "\n",
      "Do NNI search on 20 best initial trees\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 1: -32411.654\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 2: -32388.810\n",
      "Iteration 10 / LogL: -32388.817 / Time: 0h:1m:9s\n",
      "Iteration 20 / LogL: -32388.914 / Time: 0h:1m:10s\n",
      "Finish initializing candidate tree set (2)\n",
      "Current best tree score: -32388.810 / CPU time: 4.318\n",
      "Number of iterations: 20\n",
      "--------------------------------------------------------------------\n",
      "|               OPTIMIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Iteration 30 / LogL: -32388.866 / Time: 0h:1m:12s (0h:0m:15s left)\n",
      "Iteration 40 / LogL: -32411.870 / Time: 0h:1m:14s (0h:0m:13s left)\n",
      "Iteration 50 / LogL: -32388.960 / Time: 0h:1m:16s (0h:0m:10s left)\n",
      "Iteration 60 / LogL: -32398.313 / Time: 0h:1m:18s (0h:0m:8s left)\n",
      "Iteration 70 / LogL: -32396.649 / Time: 0h:1m:20s (0h:0m:6s left)\n",
      "Iteration 80 / LogL: -32411.858 / Time: 0h:1m:21s (0h:0m:4s left)\n",
      "Iteration 90 / LogL: -32389.061 / Time: 0h:1m:23s (0h:0m:2s left)\n",
      "Iteration 100 / LogL: -32411.818 / Time: 0h:1m:25s (0h:0m:0s left)\n",
      "TREE SEARCH COMPLETED AFTER 103 ITERATIONS / Time: 0h:1m:26s\n",
      "\n",
      "--------------------------------------------------------------------\n",
      "|                    FINALIZING TREE SEARCH                        |\n",
      "--------------------------------------------------------------------\n",
      "Performs final model parameters optimization\n",
      "Estimate model parameters (epsilon = 0.010)\n",
      "1. Initial log-likelihood: -32388.810\n",
      "2. Current log-likelihood: -32388.798\n",
      "Optimal log-likelihood: -32388.793\n",
      "Site proportion and rates:  (0.676,0.124) (0.255,1.688) (0.069,6.995)\n",
      "Parameters optimization took 2 rounds (0.134 sec)\n",
      "BEST SCORE FOUND : -32388.793\n",
      "Total tree length: 1.217\n",
      "\n",
      "Total number of iterations: 103\n",
      "CPU time used for tree search: 76.530 sec (0h:1m:16s)\n",
      "Wall-clock time used for tree search: 19.584 sec (0h:0m:19s)\n",
      "Total CPU time used: 299.743 sec (0h:4m:59s)\n",
      "Total wall-clock time used: 86.348 sec (0h:1m:26s)\n",
      "\n",
      "Analysis results written to: \n",
      "  IQ-TREE report:                phylogenomics/model-finder/All/All.iqtree\n",
      "  Maximum-likelihood tree:       phylogenomics/model-finder/All/All.treefile\n",
      "  Likelihood distances:          phylogenomics/model-finder/All/All.mldist\n",
      "  Screen log file:               phylogenomics/model-finder/All/All.log\n",
      "\n",
      "Date and Time: Mon Mar 17 21:00:40 2025\n"
     ]
    }
   ],
   "source": [
    "! iqtree2 -m MFP -s phylogenomics/trimmed_MSAs/All/ --prefix phylogenomics/model-finder/All/All -T AUTO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get the best substitution model!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best-fit model according to BIC: cpREV+F+R3\n",
      "\n",
      "List of models sorted by BIC scores: \n",
      "\n",
      "Model                  LogL         AIC      w-AIC        AICc     w-AICc         BIC      w-BIC\n",
      "cpREV+F+R3       -32411.668   64939.336 +    0.335   64940.980 +    0.341   65307.524 +    0.923\n"
     ]
    }
   ],
   "source": [
    "! head -42 phylogenomics/model-finder/All/All.iqtree | tail -6"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a directory to store final tree files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "mkdir phylogenomics/tree/\n",
    "mkdir phylogenomics/tree/All/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run `IQ-TREE`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IQ-TREE multicore version 2.4.0 for MacOS Intel 64-bit built Feb 12 2025\n",
      "Developed by Bui Quang Minh, Nguyen Lam Tung, Olga Chernomor, Heiko Schmidt,\n",
      "Dominik Schrempf, Michael Woodhams, Ly Trong Nhan, Thomas Wong\n",
      "\n",
      "Host:    Ilas-Mac-mini.local (SSE4.2, 16 GB RAM)\n",
      "Command: iqtree2 -s phylogenomics/trimmed_MSAs/All/ -m cpREV+F+R3 --prefix phylogenomics/tree/All/All -bb 10000 -T AUTO\n",
      "Seed:    404982 (Using SPRNG - Scalable Parallel Random Number Generator)\n",
      "Time:    Mon Mar 17 21:00:50 2025\n",
      "Kernel:  SSE2 - auto-detect threads (10 CPU cores detected)\n",
      "\n",
      "Reading 13 alignment files in directory phylogenomics/trimmed_MSAs/All/\n",
      "Reading alignment file phylogenomics/trimmed_MSAs/All/trimmed_family1.seq.aln ... Fasta format detected\n",
      "Reading fasta file: done in 0.000556946 secs using 93.01% CPU\n",
      "Alignment most likely contains protein sequences\n",
      "Alignment has 19 sequences with 258 columns, 131 distinct patterns\n",
      "73 parsimony-informative, 25 singleton sites, 160 constant sites\n",
      "           Gap/Ambiguity  Composition  p-value\n",
      "Analyzing sequences: done in 3.69549e-05 secs using 416.7% CPU\n",
      "   1  NC_015789    0.39%    passed    100.00%\n",
      "   2  NC_023125    0.00%    passed    100.00%\n",
      "   3  NC_023126    0.00%    passed    100.00%\n",
      "   4  NC_023127    0.00%    passed    100.00%\n",
      "   5  NC_023128    0.00%    passed    100.00%\n",
      "   6  NC_025200    1.16%    passed    100.00%\n",
      "   7  NC_027422    0.00%    passed    100.00%\n",
      "   8  NC_031375    0.39%    passed    100.00%\n",
      "   9  NC_033907   18.22%    passed     99.95%\n",
      "  10  NC_048507    0.00%    passed    100.00%\n",
      "  11  NC_056146    0.00%    passed    100.00%\n",
      "  12  NC_056147    0.00%    passed    100.00%\n",
      "  13  NC_056148    0.39%    passed    100.00%\n",
      "  14  NC_061762    0.00%    passed    100.00%\n",
      "  15  NC_069201    0.39%    passed    100.00%\n",
      "  16  NC_070174    0.00%    passed     99.95%\n",
      "  17  NC_071210    0.00%    passed    100.00%\n",
      "  18  NC_082275    0.00%    passed    100.00%\n",
      "  19  NC_082276    0.00%    passed    100.00%\n",
      "****  TOTAL        1.10%  0 sequences failed composition chi2 test (p-value<5%; df=19)\n",
      "Reading alignment file phylogenomics/trimmed_MSAs/All/trimmed_family10.seq.aln ... Fasta format detected\n",
      "Reading fasta file: done in 3.38554e-05 secs using 661.6% CPU\n",
      "Alignment most likely contains protein sequences\n",
      "Alignment has 19 sequences with 87 columns, 34 distinct patterns\n",
      "9 parsimony-informative, 9 singleton sites, 69 constant sites\n",
      "           Gap/Ambiguity  Composition  p-value\n",
      "Analyzing sequences: done in 1.09673e-05 secs using 838.9% CPU\n",
      "   1  NC_015789    0.00%    passed    100.00%\n",
      "   2  NC_023125    0.00%    passed    100.00%\n",
      "   3  NC_023126    0.00%    passed    100.00%\n",
      "   4  NC_023127    0.00%    passed    100.00%\n",
      "   5  NC_023128    0.00%    passed    100.00%\n",
      "   6  NC_025200    0.00%    passed    100.00%\n",
      "   7  NC_027422    0.00%    passed    100.00%\n",
      "   8  NC_031375    0.00%    passed    100.00%\n",
      "   9  NC_033907    0.00%    passed    100.00%\n",
      "  10  NC_048507    0.00%    passed    100.00%\n",
      "  11  NC_056146    0.00%    passed    100.00%\n",
      "  12  NC_056147    0.00%    passed    100.00%\n",
      "  13  NC_056148    0.00%    passed    100.00%\n",
      "  14  NC_061762    0.00%    passed    100.00%\n",
      "  15  NC_069201    0.00%    passed    100.00%\n",
      "  16  NC_070174    0.00%    passed     99.96%\n",
      "  17  NC_071210    0.00%    passed    100.00%\n",
      "  18  NC_082275    0.00%    passed    100.00%\n",
      "  19  NC_082276    0.00%    passed    100.00%\n",
      "****  TOTAL        0.00%  0 sequences failed composition chi2 test (p-value<5%; df=15)\n",
      "Reading alignment file phylogenomics/trimmed_MSAs/All/trimmed_family11.seq.aln ... Fasta format detected\n",
      "Reading fasta file: done in 0.000159979 secs using 450.7% CPU\n",
      "Alignment most likely contains protein sequences\n",
      "Alignment has 19 sequences with 653 columns, 233 distinct patterns\n",
      "150 parsimony-informative, 60 singleton sites, 442 constant sites\n",
      "           Gap/Ambiguity  Composition  p-value\n",
      "Analyzing sequences: done in 5.00679e-06 secs using 559.2% CPU\n",
      "   1  NC_015789    0.00%    passed    100.00%\n",
      "   2  NC_023125    0.00%    passed    100.00%\n",
      "   3  NC_023126    0.00%    passed    100.00%\n",
      "   4  NC_023127    0.00%    passed    100.00%\n",
      "   5  NC_023128    0.00%    passed    100.00%\n",
      "   6  NC_025200    0.00%    passed    100.00%\n",
      "   7  NC_027422    0.00%    passed    100.00%\n",
      "   8  NC_031375    0.00%    passed    100.00%\n",
      "   9  NC_033907    5.51%    passed    100.00%\n",
      "  10  NC_048507    0.00%    passed    100.00%\n",
      "  11  NC_056146    0.15%    passed    100.00%\n",
      "  12  NC_056147    0.00%    passed    100.00%\n",
      "  13  NC_056148    0.00%    passed    100.00%\n",
      "  14  NC_061762    0.00%    passed    100.00%\n",
      "  15  NC_069201    0.15%    passed     99.99%\n",
      "  16  NC_070174    0.00%    passed    100.00%\n",
      "  17  NC_071210    0.00%    passed    100.00%\n",
      "  18  NC_082275    0.00%    passed    100.00%\n",
      "  19  NC_082276    0.00%    passed    100.00%\n",
      "****  TOTAL        0.31%  0 sequences failed composition chi2 test (p-value<5%; df=19)\n",
      "Reading alignment file phylogenomics/trimmed_MSAs/All/trimmed_family12.seq.aln ... Fasta format detected\n",
      "Reading fasta file: done in 4.31538e-05 secs using 906.1% CPU\n",
      "Alignment most likely contains protein sequences\n",
      "Alignment has 19 sequences with 135 columns, 70 distinct patterns\n",
      "44 parsimony-informative, 3 singleton sites, 88 constant sites\n",
      "           Gap/Ambiguity  Composition  p-value\n",
      "Analyzing sequences: done in 1.09673e-05 secs using 866.2% CPU\n",
      "   1  NC_015789    0.00%    passed     99.96%\n",
      "   2  NC_023125    0.00%    passed     99.86%\n",
      "   3  NC_023126   12.59%    passed    100.00%\n",
      "   4  NC_023127   12.59%    passed    100.00%\n",
      "   5  NC_023128   12.59%    passed    100.00%\n",
      "   6  NC_025200    0.00%    passed    100.00%\n",
      "   7  NC_027422    0.00%    passed    100.00%\n",
      "   8  NC_031375    0.00%    passed    100.00%\n",
      "   9  NC_033907    7.41%    passed     99.97%\n",
      "  10  NC_048507    0.00%    passed    100.00%\n",
      "  11  NC_056146    0.00%    passed     99.97%\n",
      "  12  NC_056147    0.00%    passed     99.97%\n",
      "  13  NC_056148    0.00%    passed     99.97%\n",
      "  14  NC_061762    0.00%    passed    100.00%\n",
      "  15  NC_069201    0.00%    passed     99.62%\n",
      "  16  NC_070174    0.00%    passed     99.99%\n",
      "  17  NC_071210    0.00%    passed     99.97%\n",
      "  18  NC_082275    0.00%    passed    100.00%\n",
      "  19  NC_082276    0.00%    passed    100.00%\n",
      "****  TOTAL        2.38%  0 sequences failed composition chi2 test (p-value<5%; df=19)\n",
      "Reading alignment file phylogenomics/trimmed_MSAs/All/trimmed_family13.seq.aln ... Fasta format detected\n",
      "Reading fasta file: done in 0.000169992 secs using 504.7% CPU\n",
      "Alignment most likely contains protein sequences\n",
      "Alignment has 19 sequences with 526 columns, 138 distinct patterns\n",
      "64 parsimony-informative, 40 singleton sites, 422 constant sites\n",
      "           Gap/Ambiguity  Composition  p-value\n",
      "Analyzing sequences: done in 1.90735e-05 secs using 529.5% CPU\n",
      "   1  NC_015789    0.00%    passed    100.00%\n",
      "   2  NC_023125   11.03%    passed    100.00%\n",
      "   3  NC_023126   11.03%    passed    100.00%\n",
      "   4  NC_023127   11.03%    passed    100.00%\n",
      "   5  NC_023128   11.03%    passed    100.00%\n",
      "   6  NC_025200    0.00%    passed    100.00%\n",
      "   7  NC_027422    0.57%    passed    100.00%\n",
      "   8  NC_031375    0.00%    passed    100.00%\n",
      "   9  NC_033907    0.00%    passed    100.00%\n",
      "  10  NC_048507   11.03%    passed    100.00%\n",
      "  11  NC_056146    0.95%    passed    100.00%\n",
      "  12  NC_056147    0.95%    passed    100.00%\n",
      "  13  NC_056148    0.95%    passed    100.00%\n",
      "  14  NC_061762    0.19%    passed    100.00%\n",
      "  15  NC_069201    0.57%    passed    100.00%\n",
      "  16  NC_070174    0.00%    passed    100.00%\n",
      "  17  NC_071210    0.95%    passed    100.00%\n",
      "  18  NC_082275   11.03%    passed    100.00%\n",
      "  19  NC_082276   11.03%    passed    100.00%\n",
      "****  TOTAL        4.33%  0 sequences failed composition chi2 test (p-value<5%; df=19)\n",
      "Reading alignment file phylogenomics/trimmed_MSAs/All/trimmed_family2.seq.aln ... Fasta format detected\n",
      "Reading fasta file: done in 7.10487e-05 secs using 567.2% CPU\n",
      "Alignment most likely contains protein sequences\n",
      "Alignment has 19 sequences with 249 columns, 91 distinct patterns\n",
      "57 parsimony-informative, 11 singleton sites, 181 constant sites\n",
      "           Gap/Ambiguity  Composition  p-value\n",
      "Analyzing sequences: done in 1.3113e-05 secs using 648.2% CPU\n",
      "   1  NC_015789    0.00%    passed    100.00%\n",
      "   2  NC_023125    0.00%    passed    100.00%\n",
      "   3  NC_023126    0.00%    passed    100.00%\n",
      "   4  NC_023127    0.00%    passed    100.00%\n",
      "   5  NC_023128    0.00%    passed    100.00%\n",
      "   6  NC_025200    0.00%    passed    100.00%\n",
      "   7  NC_027422    0.00%    passed    100.00%\n",
      "   8  NC_031375    0.00%    passed    100.00%\n",
      "   9  NC_033907    7.23%    passed    100.00%\n",
      "  10  NC_048507    0.00%    passed    100.00%\n",
      "  11  NC_056146    0.00%    passed    100.00%\n",
      "  12  NC_056147    0.00%    passed    100.00%\n",
      "  13  NC_056148    0.00%    passed    100.00%\n",
      "  14  NC_061762    0.00%    passed    100.00%\n",
      "  15  NC_069201    0.00%    passed    100.00%\n",
      "  16  NC_070174    0.00%    passed    100.00%\n",
      "  17  NC_071210    0.00%    passed    100.00%\n",
      "  18  NC_082275    0.00%    passed    100.00%\n",
      "  19  NC_082276    0.00%    passed    100.00%\n",
      "****  TOTAL        0.38%  0 sequences failed composition chi2 test (p-value<5%; df=19)\n",
      "Reading alignment file phylogenomics/trimmed_MSAs/All/trimmed_family3.seq.aln ... Fasta format detected\n",
      "Reading fasta file: done in 7.10487e-05 secs using 914.9% CPU\n",
      "Alignment most likely contains protein sequences\n",
      "Alignment has 19 sequences with 254 columns, 116 distinct patterns\n",
      "68 parsimony-informative, 31 singleton sites, 155 constant sites\n",
      "           Gap/Ambiguity  Composition  p-value\n",
      "Analyzing sequences: done in 3.8147e-06 secs using 1022% CPU\n",
      "   1  NC_015789    0.00%    passed    100.00%\n",
      "   2  NC_023125    0.00%    passed    100.00%\n",
      "   3  NC_023126    0.00%    passed    100.00%\n",
      "   4  NC_023127    0.00%    passed    100.00%\n",
      "   5  NC_023128    0.00%    passed    100.00%\n",
      "   6  NC_025200    0.00%    passed     99.85%\n",
      "   7  NC_027422    0.00%    passed    100.00%\n",
      "   8  NC_031375    0.00%    passed    100.00%\n",
      "   9  NC_033907    0.00%    passed    100.00%\n",
      "  10  NC_048507    0.00%    passed     99.35%\n",
      "  11  NC_056146    0.00%    passed    100.00%\n",
      "  12  NC_056147    0.00%    passed    100.00%\n",
      "  13  NC_056148    0.00%    passed    100.00%\n",
      "  14  NC_061762    0.00%    passed     99.97%\n",
      "  15  NC_069201    0.00%    passed    100.00%\n",
      "  16  NC_070174    0.00%    passed    100.00%\n",
      "  17  NC_071210    0.00%    passed    100.00%\n",
      "  18  NC_082275    0.00%    passed     98.76%\n",
      "  19  NC_082276    0.00%    passed     99.35%\n",
      "****  TOTAL        0.00%  0 sequences failed composition chi2 test (p-value<5%; df=19)\n",
      "Reading alignment file phylogenomics/trimmed_MSAs/All/trimmed_family4.seq.aln ... Fasta format detected\n",
      "Reading fasta file: done in 0.000103951 secs using 618.6% CPU\n",
      "Alignment most likely contains protein sequences\n",
      "Alignment has 19 sequences with 388 columns, 134 distinct patterns\n",
      "81 parsimony-informative, 36 singleton sites, 271 constant sites\n",
      "           Gap/Ambiguity  Composition  p-value\n",
      "Analyzing sequences: done in 1.69277e-05 secs using 762.1% CPU\n",
      "   1  NC_015789    0.00%    passed    100.00%\n",
      "   2  NC_023125    0.00%    passed    100.00%\n",
      "   3  NC_023126    0.00%    passed    100.00%\n",
      "   4  NC_023127    0.52%    passed    100.00%\n",
      "   5  NC_023128    0.00%    passed    100.00%\n",
      "   6  NC_025200    0.00%    passed    100.00%\n",
      "   7  NC_027422    0.00%    passed    100.00%\n",
      "   8  NC_031375    0.00%    passed    100.00%\n",
      "   9  NC_033907    0.52%    passed    100.00%\n",
      "  10  NC_048507    0.00%    passed    100.00%\n",
      "  11  NC_056146    0.00%    passed    100.00%\n",
      "  12  NC_056147    0.00%    passed    100.00%\n",
      "  13  NC_056148    0.00%    passed    100.00%\n",
      "  14  NC_061762    0.00%    passed    100.00%\n",
      "  15  NC_069201    0.00%    passed    100.00%\n",
      "  16  NC_070174    0.00%    passed    100.00%\n",
      "  17  NC_071210    0.00%    passed    100.00%\n",
      "  18  NC_082275    0.00%    passed    100.00%\n",
      "  19  NC_082276    0.00%    passed    100.00%\n",
      "****  TOTAL        0.05%  0 sequences failed composition chi2 test (p-value<5%; df=19)\n",
      "Reading alignment file phylogenomics/trimmed_MSAs/All/trimmed_family5.seq.aln ... Fasta format detected\n",
      "Reading fasta file: done in 8.89301e-05 secs using 833.2% CPU\n",
      "Alignment most likely contains protein sequences\n",
      "Alignment has 19 sequences with 360 columns, 162 distinct patterns\n",
      "92 parsimony-informative, 46 singleton sites, 222 constant sites\n",
      "           Gap/Ambiguity  Composition  p-value\n",
      "Analyzing sequences: done in 1.5974e-05 secs using 763.7% CPU\n",
      "   1  NC_015789    0.00%    passed    100.00%\n",
      "   2  NC_023125    0.00%    passed    100.00%\n",
      "   3  NC_023126    0.00%    passed    100.00%\n",
      "   4  NC_023127    0.00%    passed    100.00%\n",
      "   5  NC_023128    0.00%    passed    100.00%\n",
      "   6  NC_025200    0.00%    passed     99.95%\n",
      "   7  NC_027422    0.00%    passed    100.00%\n",
      "   8  NC_031375    0.00%    passed    100.00%\n",
      "   9  NC_033907    1.67%    passed    100.00%\n",
      "  10  NC_048507    0.00%    passed    100.00%\n",
      "  11  NC_056146    0.00%    passed    100.00%\n",
      "  12  NC_056147    0.00%    passed    100.00%\n",
      "  13  NC_056148    0.28%    passed     99.98%\n",
      "  14  NC_061762    0.00%    passed     99.97%\n",
      "  15  NC_069201    0.00%    passed    100.00%\n",
      "  16  NC_070174    0.00%    passed    100.00%\n",
      "  17  NC_071210    0.00%    passed    100.00%\n",
      "  18  NC_082275    0.00%    passed    100.00%\n",
      "  19  NC_082276    0.00%    passed    100.00%\n",
      "****  TOTAL        0.10%  0 sequences failed composition chi2 test (p-value<5%; df=19)\n",
      "Reading alignment file phylogenomics/trimmed_MSAs/All/trimmed_family6.seq.aln ... Fasta format detected\n",
      "Reading fasta file: done in 6.91414e-05 secs using 666.7% CPU\n",
      "Alignment most likely contains protein sequences\n",
      "Alignment has 19 sequences with 223 columns, 165 distinct patterns\n",
      "128 parsimony-informative, 16 singleton sites, 79 constant sites\n",
      "           Gap/Ambiguity  Composition  p-value\n",
      "Analyzing sequences: done in 1.28746e-05 secs using 908.8% CPU\n",
      "   1  NC_015789    0.00%    passed     99.44%\n",
      "   2  NC_023125    0.00%    passed    100.00%\n",
      "   3  NC_023126    0.00%    passed    100.00%\n",
      "   4  NC_023127    0.00%    passed    100.00%\n",
      "   5  NC_023128    0.00%    passed    100.00%\n",
      "   6  NC_025200    1.79%    passed     98.88%\n",
      "   7  NC_027422    5.83%    passed     89.07%\n",
      "   8  NC_031375    2.24%    passed     99.89%\n",
      "   9  NC_033907   15.25%    passed     90.60%\n",
      "  10  NC_048507    2.24%    passed     99.66%\n",
      "  11  NC_056146    0.45%    passed     99.80%\n",
      "  12  NC_056147    0.45%    passed     99.80%\n",
      "  13  NC_056148    0.45%    passed     99.98%\n",
      "  14  NC_061762    1.35%    passed     94.09%\n",
      "  15  NC_069201   16.14%    passed     72.40%\n",
      "  16  NC_070174    2.69%    passed     97.40%\n",
      "  17  NC_071210    0.45%    passed     99.75%\n",
      "  18  NC_082275    2.24%    passed     99.05%\n",
      "  19  NC_082276    2.24%    passed     99.66%\n",
      "****  TOTAL        2.83%  0 sequences failed composition chi2 test (p-value<5%; df=19)\n",
      "Reading alignment file phylogenomics/trimmed_MSAs/All/trimmed_family7.seq.aln ... Fasta format detected\n",
      "Reading fasta file: done in 2.19345e-05 secs using 661.1% CPU\n",
      "Alignment most likely contains protein sequences\n",
      "Alignment has 19 sequences with 48 columns, 37 distinct patterns\n",
      "15 parsimony-informative, 11 singleton sites, 22 constant sites\n",
      "           Gap/Ambiguity  Composition  p-value\n",
      "Analyzing sequences: done in 3.8147e-06 secs using 104.9% CPU\n",
      "   1  NC_015789    0.00%    passed     99.99%\n",
      "   2  NC_023125    0.00%    passed     99.98%\n",
      "   3  NC_023126    0.00%    passed     99.98%\n",
      "   4  NC_023127    0.00%    passed     99.97%\n",
      "   5  NC_023128    0.00%    passed     99.98%\n",
      "   6  NC_025200    0.00%    passed    100.00%\n",
      "   7  NC_027422    0.00%    passed    100.00%\n",
      "   8  NC_031375    0.00%    passed     99.97%\n",
      "   9  NC_033907    0.00%    passed    100.00%\n",
      "  10  NC_048507    0.00%    passed    100.00%\n",
      "  11  NC_056146    0.00%    passed    100.00%\n",
      "  12  NC_056147    0.00%    passed    100.00%\n",
      "  13  NC_056148    0.00%    passed     99.95%\n",
      "  14  NC_061762    0.00%    passed    100.00%\n",
      "  15  NC_069201    0.00%    passed     13.95%\n",
      "  16  NC_070174    0.00%    passed     99.97%\n",
      "  17  NC_071210    0.00%    passed    100.00%\n",
      "  18  NC_082275    0.00%    passed     99.97%\n",
      "  19  NC_082276    0.00%    passed    100.00%\n",
      "****  TOTAL        0.00%  0 sequences failed composition chi2 test (p-value<5%; df=14)\n",
      "Reading alignment file phylogenomics/trimmed_MSAs/All/trimmed_family8.seq.aln ... Fasta format detected\n",
      "Reading fasta file: done in 0.000141144 secs using 777.9% CPU\n",
      "Alignment most likely contains protein sequences\n",
      "Alignment has 19 sequences with 552 columns, 326 distinct patterns\n",
      "231 parsimony-informative, 68 singleton sites, 253 constant sites\n",
      "           Gap/Ambiguity  Composition  p-value\n",
      "Analyzing sequences: done in 2.09808e-05 secs using 572% CPU\n",
      "   1  NC_015789    0.00%    passed     99.75%\n",
      "   2  NC_023125    0.00%    passed     99.97%\n",
      "   3  NC_023126    0.00%    passed     99.97%\n",
      "   4  NC_023127    0.00%    passed     99.95%\n",
      "   5  NC_023128    0.00%    passed     99.97%\n",
      "   6  NC_025200    0.00%    passed     99.97%\n",
      "   7  NC_027422    0.00%    passed     99.53%\n",
      "   8  NC_031375    0.00%    passed     99.93%\n",
      "   9  NC_033907   22.46%    passed     94.84%\n",
      "  10  NC_048507    0.00%    passed     99.98%\n",
      "  11  NC_056146    0.00%    passed     99.47%\n",
      "  12  NC_056147    0.36%    passed     98.32%\n",
      "  13  NC_056148    0.18%    passed     98.04%\n",
      "  14  NC_061762    0.00%    passed     99.94%\n",
      "  15  NC_069201    0.36%    passed     85.13%\n",
      "  16  NC_070174   14.31%    passed     96.04%\n",
      "  17  NC_071210    0.36%    passed     98.54%\n",
      "  18  NC_082275    0.00%    passed     99.99%\n",
      "  19  NC_082276    0.00%    passed     99.98%\n",
      "****  TOTAL        2.00%  0 sequences failed composition chi2 test (p-value<5%; df=19)\n",
      "Reading alignment file phylogenomics/trimmed_MSAs/All/trimmed_family9.seq.aln ... Fasta format detected\n",
      "Reading fasta file: done in 0.000197887 secs using 499.3% CPU\n",
      "Alignment most likely contains protein sequences\n",
      "Alignment has 19 sequences with 489 columns, 214 distinct patterns\n",
      "135 parsimony-informative, 58 singleton sites, 296 constant sites\n",
      "           Gap/Ambiguity  Composition  p-value\n",
      "Analyzing sequences: done in 1.5974e-05 secs using 770% CPU\n",
      "   1  NC_015789    0.41%    passed    100.00%\n",
      "   2  NC_023125    0.41%    passed    100.00%\n",
      "   3  NC_023126    0.41%    passed    100.00%\n",
      "   4  NC_023127    0.41%    passed    100.00%\n",
      "   5  NC_023128    0.41%    passed    100.00%\n",
      "   6  NC_025200    0.00%    passed     99.65%\n",
      "   7  NC_027422    0.00%    passed     99.92%\n",
      "   8  NC_031375    0.41%    passed     99.98%\n",
      "   9  NC_033907   15.75%    passed     99.97%\n",
      "  10  NC_048507    0.00%    passed     99.99%\n",
      "  11  NC_056146    1.84%    passed     99.97%\n",
      "  12  NC_056147    1.84%    passed     99.97%\n",
      "  13  NC_056148    1.84%    passed     99.99%\n",
      "  14  NC_061762    0.00%    passed    100.00%\n",
      "  15  NC_069201    0.20%    passed     99.74%\n",
      "  16  NC_070174    0.00%    passed     99.99%\n",
      "  17  NC_071210    1.02%    passed     99.97%\n",
      "  18  NC_082275    0.00%    passed     99.99%\n",
      "  19  NC_082276    0.00%    passed     99.99%\n",
      "****  TOTAL        1.31%  0 sequences failed composition chi2 test (p-value<5%; df=19)\n",
      "NOTE: NC_082276 is identical to NC_048507 but kept for subsequent analysis\n",
      "\n",
      "Create initial parsimony tree by phylogenetic likelihood library (PLL)... 0.002 seconds\n",
      "Generating 10000 samples for ultrafast bootstrap (seed: 404982)...\n",
      "WARNING: Default model LG may be under-fitting. Use option '-m TEST' to determine the best-fit model.\n",
      "\n",
      "NOTE: 60 MB RAM (0 GB) is required!\n",
      "Measuring multi-threading efficiency up to 10 CPU cores\n",
      "Increase to 10 rounds for branch lengths\n",
      "187 trees examined\n",
      "Threads: 1 / Time: 10.017 sec / Speedup: 1.000 / Efficiency: 100% / LogL: -52425\n",
      "Threads: 2 / Time: 6.516 sec / Speedup: 1.537 / Efficiency: 77% / LogL: -52425\n",
      "Threads: 3 / Time: 4.784 sec / Speedup: 2.094 / Efficiency: 70% / LogL: -52425\n",
      "Threads: 4 / Time: 4.224 sec / Speedup: 2.372 / Efficiency: 59% / LogL: -52425\n",
      "Threads: 5 / Time: 4.299 sec / Speedup: 2.330 / Efficiency: 47% / LogL: -52425\n",
      "BEST NUMBER OF THREADS: 4\n",
      "\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "1. Initial log-likelihood: -35153.225\n",
      "Optimal log-likelihood: -35050.531\n",
      "Parameters optimization took 1 rounds (0.010 sec)\n",
      "Wrote distance file to... \n",
      "Computing ML distances based on estimated model parameters...\n",
      "Calculating distance matrix: done in 0.00258183 secs using 444.9% CPU\n",
      "Computing ML distances took 0.002670 sec (of wall-clock time) 0.011778 sec (of CPU time)\n",
      "Setting up auxiliary I and S matrices: done in 4.19617e-05 secs using 395.6% CPU\n",
      "Computing RapidNJ tree took 0.000382 sec (of wall-clock time) 0.001508 sec (of CPU time)\n",
      "Log-likelihood of RapidNJ tree: -35129.284\n",
      "--------------------------------------------------------------------\n",
      "|             INITIALIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Generating 98 parsimony trees... 0.212 second\n",
      "Computing log-likelihood of 98 initial trees ... 0.371 seconds\n",
      "Current best score: -35050.531\n",
      "\n",
      "Do NNI search on 20 best initial trees\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 1: -35041.664\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 5: -35016.642\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "UPDATE BEST LOG-LIKELIHOOD: -35016.628\n",
      "Iteration 10 / LogL: -35041.664 / Time: 0h:0m:31s\n",
      "Iteration 20 / LogL: -35041.696 / Time: 0h:0m:33s\n",
      "Finish initializing candidate tree set (3)\n",
      "Current best tree score: -35016.628 / CPU time: 2.853\n",
      "Number of iterations: 20\n",
      "--------------------------------------------------------------------\n",
      "|               OPTIMIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Iteration 30 / LogL: -35016.631 / Time: 0h:0m:34s (0h:1m:28s left)\n",
      "Iteration 40 / LogL: -35016.630 / Time: 0h:0m:35s (0h:0m:58s left)\n",
      "Iteration 50 / LogL: -35791.456 / Time: 0h:0m:36s (0h:0m:40s left)\n",
      "Log-likelihood cutoff on original alignment: -35121.355\n",
      "Iteration 60 / LogL: -35041.663 / Time: 0h:0m:37s (0h:0m:28s left)\n",
      "Iteration 70 / LogL: -35041.665 / Time: 0h:0m:38s (0h:0m:19s left)\n",
      "Iteration 80 / LogL: -35041.665 / Time: 0h:0m:39s (0h:0m:12s left)\n",
      "Iteration 90 / LogL: -35168.634 / Time: 0h:0m:40s (0h:0m:6s left)\n",
      "Iteration 100 / LogL: -35041.664 / Time: 0h:0m:41s (0h:0m:2s left)\n",
      "Log-likelihood cutoff on original alignment: -35120.899\n",
      "NOTE: Bootstrap correlation coefficient of split occurrence frequencies: 1.000\n",
      "TREE SEARCH COMPLETED AFTER 106 ITERATIONS / Time: 0h:0m:42s\n",
      "\n",
      "--------------------------------------------------------------------\n",
      "|                    FINALIZING TREE SEARCH                        |\n",
      "--------------------------------------------------------------------\n",
      "Performs final model parameters optimization\n",
      "Estimate model parameters (epsilon = 0.010)\n",
      "1. Initial log-likelihood: -35016.628\n",
      "Optimal log-likelihood: -35016.628\n",
      "Parameters optimization took 1 rounds (0.014 sec)\n",
      "BEST SCORE FOUND : -35016.628\n",
      "Creating bootstrap support values...\n",
      "Split supports printed to NEXUS file phylogenomics/tree/All/All.splits.nex\n",
      "Total tree length: 0.961\n",
      "\n",
      "Total number of iterations: 106\n",
      "CPU time used for tree search: 45.043 sec (0h:0m:45s)\n",
      "Wall-clock time used for tree search: 11.920 sec (0h:0m:11s)\n",
      "Total CPU time used: 120.033 sec (0h:2m:0s)\n",
      "Total wall-clock time used: 42.417 sec (0h:0m:42s)\n",
      "\n",
      "Computing bootstrap consensus tree...\n",
      "Reading input file phylogenomics/tree/All/All.splits.nex...\n",
      "19 taxa and 74 splits.\n",
      "Consensus tree written to phylogenomics/tree/All/All.contree\n",
      "Reading input trees file phylogenomics/tree/All/All.contree\n",
      "Log-likelihood of consensus tree: -35016.658\n",
      "\n",
      "Analysis results written to: \n",
      "  IQ-TREE report:                phylogenomics/tree/All/All.iqtree\n",
      "  Maximum-likelihood tree:       phylogenomics/tree/All/All.treefile\n",
      "  Likelihood distances:          phylogenomics/tree/All/All.mldist\n",
      "\n",
      "Ultrafast bootstrap approximation results written to:\n",
      "  Split support values:          phylogenomics/tree/All/All.splits.nex\n",
      "  Consensus tree:                phylogenomics/tree/All/All.contree\n",
      "  Screen log file:               phylogenomics/tree/All/All.log\n",
      "\n",
      "Date and Time: Mon Mar 17 21:01:32 2025\n"
     ]
    }
   ],
   "source": [
    "! iqtree2 -s phylogenomics/trimmed_MSAs/All/ -m cpREV+F+R3 --prefix phylogenomics/tree/All/All -bb 10000 -T AUTO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now add `.1` to the accession numbers on the tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "def modify_tree_file(input_file, output_file):\n",
    "    \"\"\"\n",
    "    Reads a Newick tree from a file, adds \".1\" to all accession numbers, \n",
    "    and writes the modified tree to an output file.\n",
    "\n",
    "    Args:\n",
    "        input_file (str): Path to the input tree file.\n",
    "        output_file (str): Path to save the modified tree.\n",
    "    \"\"\"\n",
    "\n",
    "    def add_suffix(match):\n",
    "        return match.group(0) + \".1\"\n",
    "\n",
    "    # Read tree from file\n",
    "    with open(input_file, \"r\") as infile:\n",
    "        tree_str = infile.read().strip()\n",
    "\n",
    "    # Modify tree\n",
    "    modified_tree = re.sub(r'NC_\\d+', add_suffix, tree_str)\n",
    "\n",
    "    # Write modified tree to output file\n",
    "    with open(output_file, \"w\") as outfile:\n",
    "        outfile.write(modified_tree)\n",
    "\n",
    "    print(f\"Modified tree saved to: {output_file}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modified tree saved to: phylogenomics/tree/All/All_ready.treefile\n"
     ]
    }
   ],
   "source": [
    "input_tree_file = \"phylogenomics/tree/All/All.treefile\"  # Replace with actual file path\n",
    "output_tree_file = \"phylogenomics/tree/All/All_ready.treefile\"\n",
    "\n",
    "modify_tree_file(input_tree_file, output_tree_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now please proceed to the `04_ggtree_journal.R` to visualize the phylogenetic trees"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "panphylo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
